---
title: "lep-MAP3-varroa"
author: "Nurit Eliash"
date: '2022-07-05'
output:
  html_document:
    code_folding: hide
    theme: cerulean
    toc: yes
    toc_depth: 5
    toc_float:
      collapsed: yes
      df_print: paged
---
<style type="text/css">
  body{
  font-size: 12pt;
}
</style>

## AIM: Estimating recombintaion rate by constructing a linkage map for male and female varroa mites

The recombination frequency of *Varroa destructor*, a parasitic mite of honeybees was estimated for male and female adult mites. We used two analysis methods: manual calculation of exact recombination frequency, and computational estimation using a linkage mapping software, Lep-MAP3 (Rastas 2017). For both analyses we used as input a VCF file containing only the ‘Informative sites’. Informative sites are sites that are heterozygotic in the F1 female, and homozygotic for one allele in the F1 male, and his mother (F0 female). Only for these sites we can phase (determine the allele parental origin) the F2 generation genotypes, and follow the inheritance of specific sites through the generations (Fig __).  
All biosamples are available in Sequence Read Archive (SRA) under the accession [PRJNA794941](https://www.ncbi.nlm.nih.gov/bioproject/PRJNA794941/).  

## Load libraries
```{r setup }
library("tidyverse")
library("plyr")
library("dplyr")
library("ggplot2")
library("scales")
library("ggpubr")
library("gridExtra") # for arranging a few plots in one area
#library("grid")
#library("GGally")
library("data.table")
library("stringr")
library("janitor")
library("plotly") # to identify outliers. 
library("readr") # to extract numbers from a vector
library("tidyr")
library("LinkageMapView") # for constructing the linkage map
library("vcfR") # for extracting genotype data from a vcf file

knitr::opts_chunk$set(include = FALSE, eval = FALSE)
```

## Load Variant Call Format (VCF) file.

Extract genotypes for each site and individual. The metadata for all
samples can be found in
[here](https://github.com/nurit-eliash/varroa-linkage-map/blob/main/data/meta_data_223.xlsx).

```{r}
vcf <- read.vcfR("/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/data/vcf_filter/Q40BIALLDP16HDP40mis.5Chr7/Q40BIALLDP16HDP40mis.5Chr7.recode.vcf", verbose = FALSE )
vcf
# extract the genotype for each site in each individual
gt <- extract.gt(vcf, element = "GT") 
gt <- as.data.frame(t(gt)) %>%
  rownames_to_column("ID")
#clean the ID names
gt$ID <- sub("_[^_]+$", "", gt$ID)
```

We used the VCF file in two methods to estimate varroa mite recombination frequency:  
- **Lep-MAP3, a linkage mapping software (Rastas 2017).**  
- **Manual estimation of the recombination rate**, R script can be found in the file [manually-estimating-recom.Rmd](https://github.com/nurit-eliash/varroa-linkage-map/blob/main/R_scripts/manually-estimating-recom.Rmd)

# Estimating recombination frequency using Lep-MAP3
We followed the documentation in the [Lep-MAP3 Wiki page](https://sourceforge.net/p/lep-map3/wiki/LM3%20Home/), with a few adjustments (See flowchart in Fig 1).  

![](/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/varroa_lepmap/lep-map flowchart.png)

***

We ran the first two modules: **ParentCall2**, **Filtering2**, then skipped to the last module **OrderMarkers2** that produces the final map file.  
The input and output files for each modules are as follow:   
(1) **ParentCall2** module; options = removNonInformative; input = filtered VCF file and pedigree.txt; output =  data.call.  
(2) **Filtering2** module; options = dataTolerance=0.0001, removeNonInformative; input = data.call; output =  data_f.call.
(3) **OrderMarkers2** module; options = useKosambi=1 numMergeIterations=100 sexAveraged=1 outputPhasedData=2 grandparentPhase=1; input = data_f.call and map.txt files; output = order.txt and order.mapped files. 

In the following codes we generated two files necessary for the Lep-MAP analysis:  
- the list of informative sites (used to filter the original VCF file), for the first module, **ParentCall2**; 
- and the map.txt file, containing the physical position of sites, for the last module, **OrderMarkers2**.    

**Keep:**

-   Informative sites.  
-   121 adult mite samples.  

**Change:**  

-   Sample ID name (F0, F1_fem, and F1_male)  
-   site genotype for the above 3 samples, as per the cross: 

**For 0/0 x 0/1 cross:**  
    F1_fem = 0/1:22:10,12:10:353:12:444:-33.6837,0,-25.484  
    F1_male = 0/0:17:17,0:17:565:0:0:0,-5.11751,-51.1547  
    F0 = 0/0:17:17,0:17:565:0:0:0,-5.11751,-51.1547

**For 1/1 x 0/1 cross:**  
F1_fem = 0/1:22:10,12:10:353:12:444:-33.6837,0,-25.484  
F1_male = 1/1:17:0,17:0:0:17:629:-56.9466,-5.11751,0  
F0 = 1/1:17:0,17:0:0:17:629:-56.9466,-5.11751,0 

Code for changing sample genotype, see this [bash script](https://github.com/nurit-eliash/varroa-linkage-map/blob/main/data/formatVCF.sh)

## (1) Keep informative sites:

filter for informative sites in the original VCF file. 

### cross (0/0 x 0/1) 
in each family, filter for sites with:   
(a) homo x hetero F1 cross (0/0 x 0/1), so we can predict the parental
and recombinant F2 types.  
(b) homo F0 (0/0), so we can phase the hetero F1 sites.  

```{r}
df  <- setNames(data.frame(t(gt[,-1])), gt[,1]) %>% rownames_to_column("site")

# make a vector of all 30 families
families = stringr::str_extract(gt$ID, "[^_]+") %>% unique()

# make an empty list
sites = list()

for (fam in families) {
  sites[[fam]] <- df %>%
    dplyr::select(starts_with(c("site",fam))) %>%
    dplyr::filter_at(vars(matches("_dat")), all_vars(. == "0/1")) %>% 
    dplyr::filter_at(vars(matches("_son")), all_vars(. == "0/0")) %>% 
    dplyr::filter_at(vars(matches("fnd")), all_vars(. == "0/0")) 
}

# make a data frame of all sites from all families
InfoSites_00_01 <- reduce(sites, bind_rows) %>% as.data.frame() %>% select(site) %>% distinct()

InfoSites_00_01 <- separate(InfoSites_00_01, site, into = c("NW","Chr","Pos"), sep = '_') 

InfoSites_00_01$CHR <- paste(InfoSites_00_01$NW, InfoSites_00_01$Chr, sep = "_")
InfoSites_00_01 <- InfoSites_00_01 %>% select(c("CHR", "Pos"))
head(InfoSites_00_01)
```

There are 13,651 informative sites for the 0/0 x 0/1 F1 "family".  
Save the file as tsv, so it can be used in the next steps. 
```{r eval=FALSE, include=T}
#write_delim(InfoSites_00_01, "/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/results/InfoSites_00_01.tsv", col_names = FALSE)

# check manually a few IDs and sites,
#InfoSites_00_01 <- reduce(sites, bind_rows) %>% as.data.frame() %>% distinct(site, .keep_all = TRUE)

#tableInfo = InfoSites_00_01 %>% select(c(site,`57_58b_grnson`)) 
#tableAll = df %>% select(c(site,`57_58b_grnson`))
#left_join(tableInfo, tableAll, by = "site") %>% rename(c("57_58b_grnson.x" ="InfoSites", "57_58b_grnson.y" = "AllSites")) %>% view()
```

### cross (1/1 x 0/1)
```{r}
df  <- setNames(data.frame(t(gt[,-1])), gt[,1]) %>% rownames_to_column("site")

# make a vector of all 30 families
families = stringr::str_extract(gt$ID, "[^_]+") %>% unique()

# make an empty list
sites = list()
  
for (fam in families) {
  sites[[fam]] <- df %>%
    dplyr::select(starts_with(c("site",fam))) %>%
    dplyr::filter_at(vars(matches("_dat")), all_vars(. == "0/1")) %>% 
    dplyr::filter_at(vars(matches("_son")), all_vars(. == "1/1")) %>% 
    dplyr::filter_at(vars(matches("fnd")), all_vars(. == "1/1")) 
}

# make a data frame of all sites from all families
InfoSites_11_01 <- reduce(sites, bind_rows) %>% as.data.frame() %>% select(site) %>% distinct()

# make a data frame of all sites from all families
InfoSites_11_01 <- reduce(sites, bind_rows) %>% as.data.frame() %>% select(site) %>% distinct()
InfoSites_11_01 <- separate(InfoSites_11_01, site, into = c("NW","Chr","Pos"), sep = '_') 

InfoSites_11_01$CHR <- paste(InfoSites_11_01$NW, InfoSites_11_01$Chr, sep = "_")
InfoSites_11_01 <- InfoSites_11_01 %>% select(c("CHR", "Pos"))

head(InfoSites_11_01)
```
There are 12,158 informative sites for the 1/1 x 0/1 F1 "family"  
Save the file as tsv, so it can be used in the next steps. 
```{r eval=FALSE, include=T}
#write_delim(InfoSites_11_01, "/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/results/InfoSites_11_01.tsv", col_names = FALSE)

# check manually a few IDs and sites,
#InfoSites_11_01 <- reduce(sites, bind_rows) %>% as.data.frame() %>% distinct(site, .keep_all = TRUE)

#tableInfo = InfoSites_11_01 %>% select(c(site,`300_301a_grnson`)) 
#tableAll = df %>% select(c(site,`300_301a_grnson`))
#check <- left_join(tableInfo, tableAll, by = "site") %>% rename(c("300_301a_grnson.x" ="InfoSites", "300_301a_grnson.y" = "AllSites")) %>% na.omit("InfoSites")  

#all(check$InfoSites == check$AllSites)
```

## (2) Create a map.txt file with sites' physical position, for the OrderMarkers2 module

make a map.txt file with the sites assigned to 7 LGs based on their
physical position use this file as an input for the Order module.

the ordering doesn't have to be fantastic either at the end. We're just
looking to see if we can get some sort of recombination rate estimates
and a sense of which sex is showing recombination.

### cross (0/0 x 0/1)
```{r}
snps <- read_tsv("/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/varroa_lepmap/data/Q40BIALLDP16HDP40mis.5Chr7.infoSites/Q40BIALLDP16HDP40mis.5Chr7.Sites_00_01/snps_611.txt")

CHR = snps$CHR %>% unique()
LG = as.character(c(1:7))
LG_Chr = tibble(LG=as.character(c(1:7)), CHR)

map <- left_join(snps, LG_Chr, by = "CHR") %>% 
  dplyr::select("LG") 

#write_delim(map, "/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/varroa_lepmap/data/Q40BIALLDP16HDP40mis.5Chr7.infoSites/Q40BIALLDP16HDP40mis.5Chr7.Sites_00_01/map_611.txt")
```

### cross (1/1 x 0/1)
```{r}
snps <- read_tsv("/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/varroa_lepmap/data/Q40BIALLDP16HDP40mis.5Chr7.infoSites/Q40BIALLDP16HDP40mis.5Chr7.Sites_11_01/snps_202.txt")

CHR = snps$CHR %>% unique()
LG = as.character(c(1:7))
LG_Chr = tibble(LG=as.character(c(1:7)), CHR)

map <- left_join(snps, LG_Chr, by = "CHR") %>% 
  dplyr::select("LG") 

#write_delim(map, "/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/varroa_lepmap/data/Q40BIALLDP16HDP40mis.5Chr7.infoSites/Q40BIALLDP16HDP40mis.5Chr7.Sites_11_01/map_202.txt")
```

the map.txt files are used along with the Data_f.call (generated by the **Filtering2** module) will be used as input for the final module, **OrderMarkers2** module.  

the outputs of the **OrderMarkers2** module are order.mapped files.  
These are used in the next chunk:
- correlating the physical and genetic positions of the sites.  
- making the genetic map of varroa mite.  

------------------------------------------------------------------------

**Does the markers' physical position correlates to its genetic position, as determined by recombination rate?**  

# 0/0 x 0/1 cross

## Assuming recombintaions in males and females
```{r echo=FALSE, warning=FALSE,message=FALSE}
setwd("/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/varroa_lepmap/data/Q40BIALLDP16HDP40mis.5Chr7.infoSites/Q40BIALLDP16HDP40mis.5Chr7.Sites_00_01/Recom_male_fem")

# get a vector of the input files ('order_.mapped' files in the directory)
map_files = list.files(path ="/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/varroa_lepmap/data/Q40BIALLDP16HDP40mis.5Chr7.infoSites/Q40BIALLDP16HDP40mis.5Chr7.Sites_00_01/Recom_male_fem", pattern = "\\mapped$", full.names = T)

# set a list to put the plots in
out_plotSites = list()

# set a list to put the positions in, for later linkage mapping
out_Sites = list()

for (i in map_files)  {
  mapped <- read.table(i,  header =FALSE, sep ="\t")[,1:4]
  names(mapped) = c("Chr","POS","male_position", "female_position")
  mapped <- dplyr::mutate(mapped, LG =  basename(i) %>%  readr::parse_number() %>% as.character())

# calculate the total genetic size per LG
sizeFem <- unique(mapped$female_position) %>% sum()
sizeMale <- unique(mapped$male_position) %>% sum()

df <- mapped %>% pivot_longer(cols = c(male_position,female_position), names_to = "sex", values_to = "cM") 

p_site <- ggplot(df, aes(x=POS, y=cM)) +
  geom_point() +stat_cor(method = "pearson") + geom_smooth(method='lm') +
  ylab("Genetic position (cM)") +
  xlab("Physical position") + labs(title = paste0("LG: ", mapped[1,5],", Chromosome: ", mapped[1,1]), subtitle =  paste0("total size (cM): female = ", sizeFem, "; male: ",sizeMale)) +
facet_wrap("sex")

out_plotSites[[i]] = p_site
out_Sites[[i]] = df
 } 

# plot all LGs, by sex (i split into 2 plots, caus its too big)
#grid.arrange(top= grid::textGrob("Physical and genetic position correlation, assuming recombintaion in both sex", gp=grid::gpar(fontsize=24)), out_plotSites$order_1.mapped,out_plotSites$order_2.mapped, out_plotSites$order_3.mapped,out_plotSites$order_4.mapped, out_plotSites$order_5.mapped, out_plotSites$order_6.mapped, out_plotSites$order_7.mapped) 

# combine all LGs
sites_recom <-  reduce(out_Sites, bind_rows)  %>% distinct() %>% as.data.frame()

#save the markers positions 
#write.table(sites_recom,"/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/varroa_lepmap/results/00_01_cross/sites_recom.csv", col.names=T)
```

## Assuming recombinations in *FEMALES ONLY*
```{r echo=FALSE, warning=FALSE,message=FALSE}
setwd("/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/varroa_lepmap/data/Q40BIALLDP16HDP40mis.5Chr7.infoSites/Q40BIALLDP16HDP40mis.5Chr7.Sites_00_01/0_male")

# get a vector of the input files ('order_.mapped' files in the directory)
map_files = list.files(path ="/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/varroa_lepmap/data/Q40BIALLDP16HDP40mis.5Chr7.infoSites/Q40BIALLDP16HDP40mis.5Chr7.Sites_00_01/0_male", pattern = "\\mapped$", full.names = TRUE)

# set a list to put the plots in
out_plotSites = list()

# set a list to put the positions in, for later linkage mapping
out_Sites = list()

for (i in map_files) {
  mapped <- read.table(i,  header =FALSE, sep ="\t")[,1:4]
  names(mapped) = c("Chr","POS","male_position", "female_position")
  mapped <- dplyr::mutate(mapped, LG =  basename(i) %>%  readr::parse_number() %>% as.character())

# calculate the total genetic size per LG
sizeFem <- unique(mapped$female_position) %>% sum()
sizeMale <- unique(mapped$male_position) %>% sum()

df <- mapped %>% pivot_longer(cols = c(male_position,female_position), names_to = "sex", values_to = "cM") 

p_site <- ggplot(df, aes(x=POS, y=cM)) +
  geom_point() +stat_cor(method = "pearson") + geom_smooth(method='lm') +
  ylab("Genetic position (cM)") +
  xlab("Physical position") + labs(title = paste0("LG: ", mapped[1,5],", Chromosome: ", mapped[1,1]), subtitle =  paste0("total size (cM): female = ", sizeFem, "; male: ",sizeMale)) +
facet_wrap("sex")

out_plotSites[[i]] = p_site
out_Sites[[i]] = df
 }

# plot all LGs, by sex
#grid.arrange(top= grid::textGrob("Physical and genetic position correlation, assuming recombintaion only in FEMALES", gp=grid::gpar(fontsize=24)),  out_plotSites$order_1.mapped,out_plotSites$order_2.mapped, out_plotSites$order_3.mapped,out_plotSites$order_4.mapped, out_plotSites$order_5.mapped, out_plotSites$order_6.mapped, out_plotSites$order_7.mapped) 

# combine all LGs
sites_0male <-  reduce(out_Sites, bind_rows)  %>% distinct() %>% as.data.frame()

#save the markers positions 
#write.table(sites_0male,"/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/varroa_lepmap/results/00_01_cross/sites_0male.csv", col.names=T)
```

## Assuming recombintaions in *MALES only*
```{r echo=FALSE, warning=FALSE,message=FALSE}
setwd("/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/varroa_lepmap/data/Q40BIALLDP16HDP40mis.5Chr7.infoSites/Q40BIALLDP16HDP40mis.5Chr7.Sites_00_01/0_fem")

# get a vector of the input files ('order_.mapped' files in the directory)
map_files = list.files(path ="/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/varroa_lepmap/data/Q40BIALLDP16HDP40mis.5Chr7.infoSites/Q40BIALLDP16HDP40mis.5Chr7.Sites_00_01/0_fem", pattern = "\\mapped$", full.names = TRUE)

# set a list to put the plots in
out_plotSites = list()

# set a list to put the positions in, for later linkage mapping
out_Sites = list()

for (i in map_files) {
  mapped <- read.table(i,  header =FALSE, sep ="\t")[,1:4]
  names(mapped) = c("Chr","POS","male_position", "female_position")
  mapped <- dplyr::mutate(mapped, LG =  basename(i) %>%  readr::parse_number() %>% as.character())

# calculate the total genetic size per LG
sizeFem <- unique(mapped$female_position) %>% sum()
sizeMale <- unique(mapped$male_position) %>% sum()

df <- mapped %>% pivot_longer(cols = c(male_position,female_position), names_to = "sex", values_to = "cM") 

p_site <- ggplot(df, aes(x=POS, y=cM)) +
  geom_point() +stat_cor(method = "pearson") + geom_smooth(method='lm') +
  ylab("Genetic position (cM)") +
  xlab("Physical position") + labs(title = paste0("LG: ", mapped[1,5],", Chromosome: ", mapped[1,1]), subtitle =  paste0("total size (cM): female = ", sizeFem, "; male: ",sizeMale)) +
facet_wrap("sex")

out_plotSites[[i]] = p_site
out_Sites[[i]] = df
 }

# plot all LGs, by sex
#grid.arrange(top= grid::textGrob("Physical and genetic position correlation, assuming recombintaion only in MALES", gp=grid::gpar(fontsize=24)),  out_plotSites$order_1.mapped,out_plotSites$order_2.mapped, out_plotSites$order_3.mapped,out_plotSites$order_4.mapped, out_plotSites$order_5.mapped, out_plotSites$order_6.mapped, out_plotSites$order_7.mapped) 

# combine all LGs
sites_0fem <-  reduce(out_Sites, bind_rows)  %>% distinct() %>% as.data.frame()
#save the markers positions 
#write.table(sites_0fem,"/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/varroa_lepmap/results/00_01_cross/sites_0fem.csv", col.names=T)
```

# 1/1 x 0/1 cross

## Assuming recombintaions in males and females
```{r echo=FALSE, warning=FALSE,message=FALSE}
setwd("/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/varroa_lepmap/data/Q40BIALLDP16HDP40mis.5Chr7.infoSites/Q40BIALLDP16HDP40mis.5Chr7.Sites_11_01/Recom_male_fem")

# get a vector of the input files ('order_.mapped' files in the directory)
map_files = list.files(path ="/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/varroa_lepmap/data/Q40BIALLDP16HDP40mis.5Chr7.infoSites/Q40BIALLDP16HDP40mis.5Chr7.Sites_11_01/Recom_male_fem", pattern = "\\mapped$", full.names = TRUE)

# set a list to put the plots in
out_plotSites = list()

# set a list to put the positions in, for later linkage mapping
out_Sites = list()

for (i in map_files) {
  mapped <- read.table(i,  header =FALSE, sep ="\t")[,1:4]
  names(mapped) = c("Chr","POS","male_position", "female_position")
  mapped <- dplyr::mutate(mapped, LG =  basename(i) %>%  readr::parse_number() %>% as.character())

# calculate the total genetic size per LG
sizeFem <- unique(mapped$female_position) %>% sum()
sizeMale <- unique(mapped$male_position) %>% sum()

df <- mapped %>% pivot_longer(cols = c(male_position,female_position), names_to = "sex", values_to = "cM") 

p_site <- ggplot(df, aes(x=POS, y=cM)) +
  geom_point() +stat_cor(method = "pearson") + geom_smooth(method='lm') +
  ylab("Genetic position (cM)") +
  xlab("Physical position") + labs(title = paste0("LG: ", mapped[1,5],", Chromosome: ", mapped[1,1]), subtitle =  paste0("total size (cM): female = ", sizeFem, "; male: ",sizeMale)) +
facet_wrap("sex")

out_plotSites[[i]] = p_site
out_Sites[[i]] = df
 }

# plot all LGs, by sex (i split into 2 plots, caus its too big)
#grid.arrange(top= grid::textGrob("Physical and genetic position correlation, assuming recombintaion in both sex", gp=grid::gpar(fontsize=24)), out_plotSites$order_1.mapped,out_plotSites$order_2.mapped, out_plotSites$order_3.mapped,out_plotSites$order_4.mapped, out_plotSites$order_5.mapped, out_plotSites$order_6.mapped, out_plotSites$order_7.mapped) 

# combine all LGs
sites_recom <-  reduce(out_Sites, bind_rows)  %>% distinct() %>% as.data.frame()

#save the markers positions 
#write.table(sites_recom,"/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/varroa_lepmap/results/11_01_cross/sites_recom.csv", col.names=T)
```

## Assuming recombinations in *FEMALES ONLY*
```{r echo=FALSE, warning=FALSE,message=FALSE}
setwd("/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/varroa_lepmap/data/Q40BIALLDP16HDP40mis.5Chr7.infoSites/Q40BIALLDP16HDP40mis.5Chr7.Sites_11_01/0_male")

# get a vector of the input files ('order_.mapped' files in the directory)
map_files = list.files(path ="/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/varroa_lepmap/data/Q40BIALLDP16HDP40mis.5Chr7.infoSites/Q40BIALLDP16HDP40mis.5Chr7.Sites_11_01/0_male", pattern = "\\mapped$", full.names = TRUE)

# set a list to put the plots in
out_plotSites = list()

# set a list to put the positions in, for later linkage mapping
out_Sites = list()

for (i in map_files) {
  mapped <- read.table(i,  header =FALSE, sep ="\t")[,1:4]
  names(mapped) = c("Chr","POS","male_position", "female_position")
  mapped <- dplyr::mutate(mapped, LG =  basename(i) %>%  readr::parse_number() %>% as.character())

# calculate the total genetic size per LG
sizeFem <- unique(mapped$female_position) %>% sum()
sizeMale <- unique(mapped$male_position) %>% sum()

df <- mapped %>% pivot_longer(cols = c(male_position,female_position), names_to = "sex", values_to = "cM") 

p_site <- ggplot(df, aes(x=POS, y=cM)) +
  geom_point() +stat_cor(method = "pearson") + geom_smooth(method='lm') +
  ylab("Genetic position (cM)") +
  xlab("Physical position") + labs(title = paste0("LG: ", mapped[1,5],", Chromosome: ", mapped[1,1]), subtitle =  paste0("total size (cM): female = ", sizeFem, "; male: ",sizeMale)) +
facet_wrap("sex")

out_plotSites[[i]] = p_site
out_Sites[[i]] = df
 }

# plot all LGs, by sex
#grid.arrange(top= grid::textGrob("Physical and genetic position correlation, assuming recombintaion only in FEMALES", gp=grid::gpar(fontsize=24)),  out_plotSites$order_1.mapped,out_plotSites$order_2.mapped, out_plotSites$order_3.mapped,out_plotSites$order_4.mapped, out_plotSites$order_5.mapped, out_plotSites$order_6.mapped, out_plotSites$order_7.mapped) 

# combine all LGs
sites_0male <-  reduce(out_Sites, bind_rows)  %>% distinct() %>% as.data.frame()

#save the markers positions 
#write.table(sites_0male,"/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/varroa_lepmap/results/11_01_cross/sites_0male.csv", col.names=T)
```

## Assuming recombintaions in *MALES only*
```{r echo=FALSE, warning=FALSE,message=FALSE}
setwd("/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/varroa_lepmap/data/Q40BIALLDP16HDP40mis.5Chr7.infoSites/Q40BIALLDP16HDP40mis.5Chr7.Sites_11_01/0_fem")

# get a vector of the input files ('order_.mapped' files in the directory)
map_files = list.files(path ="/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/varroa_lepmap/data/Q40BIALLDP16HDP40mis.5Chr7.infoSites/Q40BIALLDP16HDP40mis.5Chr7.Sites_11_01/0_fem", pattern = "\\mapped$", full.names = TRUE)

# set a list to put the plots in
out_plotSites = list()

# set a list to put the positions in, for later linkage mapping
out_Sites = list()

for (i in map_files) {
  mapped <- read.table(i,  header =FALSE, sep ="\t")[,1:4]
  names(mapped) = c("Chr","POS","male_position", "female_position")
  mapped <- dplyr::mutate(mapped, LG =  basename(i) %>%  readr::parse_number() %>% as.character())

# calculate the total genetic size per LG
sizeFem <- unique(mapped$female_position) %>% sum()
sizeMale <- unique(mapped$male_position) %>% sum()

df <- mapped %>% pivot_longer(cols = c(male_position,female_position), names_to = "sex", values_to = "cM") 

p_site <- ggplot(df, aes(x=POS, y=cM)) +
  geom_point() +stat_cor(method = "pearson") + geom_smooth(method='lm') +
  ylab("Genetic position (cM)") +
  xlab("Physical position") + labs(title = paste0("LG: ", mapped[1,5],", Chromosome: ", mapped[1,1]), subtitle =  paste0("total size (cM): female = ", sizeFem, "; male: ",sizeMale)) +
facet_wrap("sex")

out_plotSites[[i]] = p_site
out_Sites[[i]] = df
 }

# plot all LGs, by sex
#grid.arrange(top= grid::textGrob("Physical and genetic position correlation, assuming recombintaion only in MALES", gp=grid::gpar(fontsize=24)),  out_plotSites$order_1.mapped,out_plotSites$order_2.mapped, out_plotSites$order_3.mapped,out_plotSites$order_4.mapped, out_plotSites$order_5.mapped, out_plotSites$order_6.mapped, out_plotSites$order_7.mapped) 

# combine all LGs
sites_0fem <-  reduce(out_Sites, bind_rows)  %>% distinct() %>% as.data.frame()
#save the markers positions 
#write.table(sites_0fem,"/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/varroa_lepmap/results/11_01_cross/sites_0fem.csv", col.names=T)
```

## Genetic map based on 00 x 01 cross

The genetic maps are based on two different crosses: 
 - 0/0 x 0/1
 - 1/1 x 0/1

### Markers positions, assuming recombinations in *both sexes*
```{r}
#load data of the markers position 
sites_recom = read.csv(file = "/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/varroa_lepmap/results/00_01_cross/sites_recom.csv", header = TRUE, sep = " ")
# ~/Documents/GitHub/varroa-linkage-map/varroa_lepmap/script/lep-MAP3-varroa.Rmd

# change the names of the columns, so it will fit for the lmv command:
#>     group position    locus
#> 1 2170LG3    0.000 BSSR-094
#> 2 2170LG3    7.039   ESSR86
#> 3 2170LG3   11.123      F3H
#> 4 2170LG3   11.123     FLS1

#"group" is the LG (1-7) + add the sex
# "position" is the genetic position (cM on the LG)
# "locus" is the physical position (Chr_pos)

sites_recom_map <- sites_recom %>% dplyr::select(c(group=LG, position = cM, locus = POS, sex = sex))
head(sites_recom_map)

## Female positions, assuming recombination in both sexes ###
cM_Recom_femPOS <- sites_recom_map %>% filter(sex=="female_position") %>% select(-sex)
head(cM_Recom_femPOS)

#outfile = file.path("/Users/nuriteliash/Desktop", "cM_Recom_femPOS.pdf")
#lmv.linkage.plot(mapthis = cM_Recom_femPOS, outfile = outfile, main="Female positions, assuming recombination in both sexes",col.main = "red")

# density map: denmap=TRUE
outfile = file.path("/Users/nuriteliash/Desktop", "cM_Recom_femPOS_dens_00_01.pdf")
lmv.linkage.plot(mapthis = cM_Recom_femPOS, outfile = outfile, main="Female positions, based on 0/0 x 0/1 cross, assuming recombination in both sexes",col.main = "red", denmap=TRUE)

## Male positions, assuming recombination in both sexes
cM_Recom_malePOS <- sites_recom_map %>% filter(sex=="male_position") %>% select(-sex)
head(cM_Recom_malePOS)
  
#outfile = file.path("/Users/nuriteliash/Desktop", "cM_Recom_malePOS.pdf")
#lmv.linkage.plot(mapthis = cM_Recom_malePOS, outfile = outfile, main="Male positions, assuming recombination in both sexes",col.main = "blue")

# density map: denmap=TRUE
outfile = file.path("/Users/nuriteliash/Desktop", "cM_Recom_malePOS_dens_00_01.pdf")
lmv.linkage.plot(mapthis = cM_Recom_malePOS, outfile = outfile, main="Male positions, based on 0/0 x 0/1 cross, assuming recombination in both sexes",col.main = "blue", denmap=TRUE)
```