---
title: "RIdeogram"
author: "Nurit Eliash"
date: '2022-07-15'
output:
  html_document:
    code_folding: hide
    theme: cerulean
    toc: yes
    toc_depth: 5
    toc_float:
      collapsed: yes
      df_print: paged
---
<style type="text/css">
  body{
  font-size: 12pt;
}
</style>

## Load libraries
```{r setup, include=FALSE, warning=FALSE}
library("RIdeogram") # for visualizing the variant sites on the chromosomes
library("tidyverse")
library("plyr")
library("dplyr")
library("ggplot2")
library("scales")
library("ggpubr")
library("gridExtra")
library("grid")
library("GGally")
library("vcfR") # for extracting genotype data from a vcf file
library("data.table")
library("stringr")
library("janitor")
library("PopGenome") # for computing the running mean, using the 'SlidingWindow' function
library("readr") # to read a delimited table - remove this and the PopGwnome if i dont use it at the end
library("runner") # to calculate running windows
library("Rcpp") # required for the runner package
knitr::opts_chunk$set(echo = TRUE)
```



```{r}
 df= read_delim("/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/varroa_lepmap/results/Chr_all.csv")

```


make a table with chromosome info, and plot their position
```{r eval=FALSE, include=FALSE}
# make the karyotype table with chromosomes start and end position
karyotype <- tibble(Chr = c("NW_019211454.1", "NW_019211455.1", "NW_019211456.1", "NW_019211457.1", "NW_019211458.1", "NW_019211459.1", "NW_019211460.1"),
           Start = 0,
             End = c(76960006, 60513814,58583513,52932055,42024542,32556157,39431147),
           ) %>% separate(col=Chr, into=c("noting","Chr"), sep="_") %>%
  dplyr::select(-"noting")
  
sites <- F2male %>%
    dplyr::select(c("Chr","Start","End","count0/1")) %>%
    dplyr::rename("Value" = "count0/1")

#ideogram(karyotype = karyotype)
ideogram(karyotype = karyotype, overlaid = sites)
convertSVG("chromosome.svg", device = "png")
```

