---
title: "Varroa genotype distribution"
author: "Nurit Eliash"
date: "May 2023"
output:
  html_document:
    code_folding: hide
    theme: cerulean
    toc: yes
    toc_depth: 5
    toc_float:
      collapsed: yes
      df_print: paged
editor_options: 
  chunk_output_type: console
---

```{=html}
<style type="text/css">
  body{
  font-size: 12pt;
}
</style>
```

## load libraries
```{r libraries, echo=TRUE, message=FALSE, warning=FALSE}
library("tidyverse")
library("plyr")
library("dplyr")
library("ggplot2")
library("scales")
library("ggpubr")
library("gridExtra")
library("grid")
library("GGally")
library("vcfR") # for extracting genotype data from a vcf file
library("data.table")
library("stringr")
library("janitor")
library("plotly") # for the 3d plots
library("ggfortify")
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE,
                     fig.width = 10,
                      fig.asp = 0.6,
                      out.width = "100%")
```

## Load Variant Call Format (VCF) file.
Extract genotypes for each site and individual. The metadata for all samples can be found in [here](https://github.com/nurit-eliash/varroa-linkage-map/blob/main/data/meta_data_223.xlsx).

```{r}
vcf <- read.vcfR("/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/data/vcf_filter/Q40BIALLDP16HDP40mis.5Chr7/Q40BIALLDP16HDP40mis.5Chr7.recode.vcf", verbose = FALSE )
vcf
# extract the genotype for each site in each individual
gt <- extract.gt(vcf, element = "GT") 
gt <- as.data.frame(t(gt)) %>%
  rownames_to_column("ID")
#clean the ID names
gt$ID <- sub("_[^_]+$", "", gt$ID)

table <-  gt %>% 
  t() %>%
  as.data.frame() %>%
  row_to_names(row_number = 1) # %>%
 # dplyr::select(-c("48_49e_grn", "48_49f_grn", "48_49g_grn")) # remove samples: 48_49e_grn, 48_49f_grn, 48_49g_grn. they were born in a second cycle, and in the first cycle an adult male was born in F2. so i cannot tell for sure who is the father 
```

read the meta data table and set the family
```{r}
meta <- read_csv("/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/data/meta_more.csv")

# set the families 
family = grep("fnd",gt$ID, value=TRUE) %>%
  str_extract("[^_]+")  %>%
  unique()
```

## Control: Homozygotic sites (crosses 1-4)
### (1) F1 male (0/0) x female (0/0)
we expect all F2 offspring to be homozygotic (0/0) like their parents (F1)
```{r}
# Prepare table with observed and expected counts, per family:
# define a list to put all the data frames in
obs <- list()

for (fam in family) {
 
obs[[fam]] <- table %>%
  dplyr::select(starts_with(fam)) %>%
        dplyr::filter_at(vars(matches("_fnd")), all_vars(. == "0/0")) %>% # force F0 female to be homo, like her son
dplyr::filter_at(vars(matches("_son")), all_vars(. == "0/0")) %>%
  dplyr::filter_at(vars(matches("_dat")), all_vars(. == "0/0")) %>% 
  dplyr::select(contains("grn")) %>%
  tidyr::pivot_longer(everything())  %>% 
  dplyr::rename(sample = name, gt = value) %>%
  dplyr::count(sample, gt, .drop = FALSE) %>% 
  dplyr::filter(gt %in% c("0/0", "1/1", "0/1")) %>%
  mutate(n = as.numeric(n)) #%>%
#  group_by(sample,gt) %>%
 # mutate(total = as.numeric(sum(n)))
}

# bind all families together, to a final data frame containing all observed counts
observed <- do.call("rbind", obs) %>% mutate(sample = as.character(sample))
# replace "NA" with zero value when gt doesn't exists
samples <- data.frame(sample = rep(unique(observed$sample), each =3), gt = rep(c("0/0", "0/1", "1/1"))) # add all possible genotypes

df <- left_join(samples, observed, by=c("sample","gt")) %>%  
    replace(is.na(.),0) %>% # replace "NA" with zero value when gt doesn't exists 
    dplyr::group_by(sample) %>% 
    dplyr::mutate(total = as.numeric(sum(n))) %>% 
    dplyr::mutate(freq=round(n/total, digit=2)) %>%
    left_join(meta, by="sample") 

# order the stages 
df$sex <- factor(df$sex, level=c("female", "male", "Nymph", "Egg"))

# wrap by gt
p = df %>% dplyr::filter(total>10) %>%
  ggplot(aes(x = sex, y = freq, colour = gt, shape = adult_sisters, 
              text = paste("sample:", sample,
                          ", n:", total))) + 
  geom_boxplot(outlier.shape = NA, coef=0 ) +
  geom_jitter(width=0.1, size=2) +
  scale_color_manual(values=c("#ffbf00", "#66b032","#1982c4")) +
  theme_bw() +
scale_shape_manual(values=c(1,19)) +
  ggtitle('F2 offspring of F1 male (0/0) x female (0/0)') +
    facet_wrap( ~gt)  +  guides(color = "none") +
 theme(plot.margin = margin(1, 1, 1, 1, "cm"))

# warp by sex
p = df %>% dplyr::filter(total>10) %>%
  ggplot(aes(x = gt, y = freq, colour = gt, shape = adult_sisters, 
              text = paste("sample:", sample,
                          ", n:", total))) + 
  geom_boxplot(outlier.shape = NAf, coef=0 ) +
  geom_jitter(width=0.1, size=2) +
  scale_color_manual(values=c("#ffbf00", "#66b032","#1982c4")) +
  theme_bw() +
scale_shape_manual(values=c(1,19)) +
  ggtitle('F2 offspring of F1 male (0/0) x female (0/0)') +
    facet_wrap( ~sex)  +  guides(color = "none") +
 theme(plot.margin = margin(1, 1, 1, 1, "cm"))
```

```{r,  echo=FALSE,  out.width="100%", out.height="50%"}
ggplotly(p) %>% 
  layout(legend = list(orientation = "h", x = 0, y = -0.1),
         title = list(text = paste0('F2 offspring of F1 male (0/0) x female (0/0), ',
                                    '<br>',
                                    paste0("site count= ", sum(df$n)))))
```

### (2) F1 male (1/1) x female (1/1)
we expect all F2 offspring to be homozygotic (1/1) like their parents (F1)
```{r}
# Prepare table with observed and expected counts, per family:
# define a list to put all the data frames in
obs <- list()

for (fam in family) {
 
obs[[fam]] <- table %>%
  dplyr::select(starts_with(fam)) %>%
        dplyr::filter_at(vars(matches("_fnd")), all_vars(. == "1/1")) %>% # force F0 female to be homo, like her son
dplyr::filter_at(vars(matches("_son")), all_vars(. == "1/1")) %>%
  dplyr::filter_at(vars(matches("_dat")), all_vars(. == "1/1")) %>% 
  dplyr::select(contains("grn")) %>%
  tidyr::pivot_longer(everything())  %>% 
  dplyr::rename(sample = name, gt = value) %>%
  dplyr::count(sample, gt, .drop = FALSE) %>% 
  dplyr::filter(gt %in% c("0/0", "1/1", "0/1")) %>%
  mutate(n = as.numeric(n)) #%>%
#  group_by(sample,gt) %>%
 # mutate(total = as.numeric(sum(n)))
}

# bind all families together, to a final data frame containing all observed counts
observed <- do.call("rbind", obs) %>% mutate(sample = as.character(sample))
# replace "NA" with zero value when gt doesn't exists
samples <- data.frame(sample = rep(unique(observed$sample), each =3), gt = rep(c("0/0", "0/1", "1/1"))) # add all possible genotypes

df <- left_join(samples, observed, by=c("sample","gt")) %>%  
    replace(is.na(.),0) %>% # replace "NA" with zero value when gt doesn't exists 
    dplyr::group_by(sample) %>% 
    dplyr::mutate(total = as.numeric(sum(n))) %>% 
    dplyr::mutate(freq=round(n/total, digit=2)) %>%
    left_join(meta, by="sample") 

# order the stages 
df$sex <- factor(df$sex, level=c("female", "male", "Nymph", "Egg"))

# wrap by gt
p = df %>% dplyr::filter(total>10) %>%
  ggplot(aes(x = sex, y = freq, colour = gt, shape = adult_sisters, 
              text = paste("sample:", sample,
                          ", n:", total))) + 
  geom_boxplot(outlier.shape = NA, coef=0 ) +
  geom_jitter(width=0.1, size=2) +
  scale_color_manual(values=c("#ffbf00", "#66b032","#1982c4")) +
  theme_bw() +
scale_shape_manual(values=c(1,19)) +
  ggtitle('F2 offspring of F1 male (1/1) x female (1/1)') +
    facet_wrap( ~gt)  +  guides(color = "none") +
 theme(plot.margin = margin(1, 1, 1, 1, "cm"))

# warp by sex
p = df %>% dplyr::filter(total>10) %>%
  ggplot(aes(x = gt, y = freq, colour = gt, shape = adult_sisters, 
              text = paste("sample:", sample,
                          ", n:", total))) + 
  geom_boxplot(outlier.shape = NAf, coef=0 ) +
  geom_jitter(width=0.1, size=2) +
  scale_color_manual(values=c("#ffbf00", "#66b032","#1982c4")) +
  theme_bw() +
scale_shape_manual(values=c(1,19)) +
  ggtitle('F2 offspring of F1 male (1/1) x female (1/1)') +
    facet_wrap( ~sex)  +  guides(color = "none") +
 theme(plot.margin = margin(1, 1, 1, 1, "cm"))
```

```{r,  echo=FALSE,  out.width="100%", out.height="50%"}
ggplotly(p) %>% 
  layout(legend = list(orientation = "h", x = 0, y = -0.1),
         title = list(text = paste0('F2 offspring of F1 male (1/1) x female (1/1), ',
                                    '<br>',
                                    paste0("site count= ", sum(df$n)))))
```

### (3) F1 male (0/0) x female (1/1)
We expect zero sites for this cross, since there is no paternal inheritance to the males.\
Indeed, there are only 25 sites in the F2 offspring, and the genotype distribution is random.  
```{r}
# Prepare table with observed and expected counts, per family:
# define a list to put all the data frames in
obs <- list()

for (fam in family) {
 
obs[[fam]] <- table %>%
  dplyr::select(starts_with(fam)) %>%
        dplyr::filter_at(vars(matches("_fnd")), all_vars(. == "0/0")) %>% # force F0 female to be homo, like her son
dplyr::filter_at(vars(matches("_son")), all_vars(. == "0/0")) %>%
  dplyr::filter_at(vars(matches("_dat")), all_vars(. == "1/1")) %>% 
  dplyr::select(contains("grn")) %>%
  tidyr::pivot_longer(everything())  %>% 
  dplyr::rename(sample = name, gt = value) %>%
  dplyr::count(sample, gt, .drop = FALSE) %>% 
  dplyr::filter(gt %in% c("0/0", "1/1", "0/1")) %>%
  mutate(n = as.numeric(n)) #%>%
#  group_by(sample,gt) %>%
 # mutate(total = as.numeric(sum(n)))
}

# bind all families together, to a final data frame containing all observed counts
observed <- do.call("rbind", obs) %>% mutate(sample = as.character(sample))
# replace "NA" with zero value when gt doesn't exists
samples <- data.frame(sample = rep(unique(observed$sample), each =3), gt = rep(c("0/0", "0/1", "1/1"))) # add all possible genotypes

df <- left_join(samples, observed, by=c("sample","gt")) %>%  
    replace(is.na(.),0) %>% # replace "NA" with zero value when gt doesn't exists 
    dplyr::group_by(sample) %>% 
    dplyr::mutate(total = as.numeric(sum(n))) %>% 
    dplyr::mutate(freq=round(n/total, digit=2)) %>%
    left_join(meta, by="sample") 

# order the stages 
df$sex <- factor(df$sex, level=c("female", "male", "Nymph", "Egg"))

# wrap by gt
p = df %>% dplyr::filter(total>10) %>%
  ggplot(aes(x = sex, y = freq, colour = gt, shape = adult_sisters, 
              text = paste("sample:", sample,
                          ", n:", total))) + 
  geom_boxplot(outlier.shape = NA, coef=0 ) +
  geom_jitter(width=0.1, size=2) +
  scale_color_manual(values=c("#ffbf00", "#66b032","#1982c4")) +
  theme_bw() +
scale_shape_manual(values=c(1,19)) +
  ggtitle('F2 offspring of F1 male (0/0) x female (1/1)') +
    facet_wrap( ~gt)  +  guides(color = "none") +
 theme(plot.margin = margin(1, 1, 1, 1, "cm"))

# warp by sex
p = df %>% dplyr::filter(total>10) %>%
  ggplot(aes(x = gt, y = freq, colour = gt, shape = adult_sisters, 
              text = paste("sample:", sample,
                          ", n:", total))) + 
  geom_boxplot(outlier.shape = NAf, coef=0 ) +
  geom_jitter(width=0.1, size=2) +
  scale_color_manual(values=c("#ffbf00", "#66b032","#1982c4")) +
  theme_bw() +
scale_shape_manual(values=c(1,19)) +
  ggtitle('F2 offspring of F1 male (0/0) x female (1/1)') +
    facet_wrap( ~sex)  +  guides(color = "none") +
 theme(plot.margin = margin(1, 1, 1, 1, "cm"))
```

```{r,  echo=FALSE,  out.width="100%", out.height="50%"}
ggplotly(p) %>% 
  layout(legend = list(orientation = "h", x = 0, y = -0.1),
         title = list(text = paste0('F2 offspring of F1 male (0/0) x female (1/1), ',
                                    '<br>',
                                    paste0("site count= ", sum(df$n)))))
```

### (4) F1 male (1/1) x female (0/0)
We expect zero sites for this cross, since there is no paternal inheritance to the males.\
Indeed, there are only 24 sites in the F2 pooled females, and 15 for the F2 pooled males:
```{r}
# Prepare table with observed and expected counts, per family:
# define a list to put all the data frames in
obs <- list()

for (fam in family) {
 
obs[[fam]] <- table %>%
  dplyr::select(starts_with(fam)) %>%
        dplyr::filter_at(vars(matches("_fnd")), all_vars(. == "1/1")) %>% # force F0 female to be homo, like her son
dplyr::filter_at(vars(matches("_son")), all_vars(. == "1/1")) %>%
  dplyr::filter_at(vars(matches("_dat")), all_vars(. == "0/0")) %>% 
  dplyr::select(contains("grn")) %>%
  tidyr::pivot_longer(everything())  %>% 
  dplyr::rename(sample = name, gt = value) %>%
  dplyr::count(sample, gt, .drop = FALSE) %>% 
  dplyr::filter(gt %in% c("0/0", "1/1", "0/1")) %>%
  mutate(n = as.numeric(n)) #%>%
#  group_by(sample,gt) %>%
 # mutate(total = as.numeric(sum(n)))
}

# bind all families together, to a final data frame containing all observed counts
observed <- do.call("rbind", obs) %>% mutate(sample = as.character(sample))
# replace "NA" with zero value when gt doesn't exists
samples <- data.frame(sample = rep(unique(observed$sample), each =3), gt = rep(c("0/0", "0/1", "1/1"))) # add all possible genotypes

df <- left_join(samples, observed, by=c("sample","gt")) %>%  
    replace(is.na(.),0) %>% # replace "NA" with zero value when gt doesn't exists 
    dplyr::group_by(sample) %>% 
    dplyr::mutate(total = as.numeric(sum(n))) %>% 
    dplyr::mutate(freq=round(n/total, digit=2)) %>%
    left_join(meta, by="sample") 

# order the stages 
df$sex <- factor(df$sex, level=c("female", "male", "Nymph", "Egg"))

# wrap by gt
p = df %>% dplyr::filter(total>10) %>%
  ggplot(aes(x = sex, y = freq, colour = gt, shape = adult_sisters, 
              text = paste("sample:", sample,
                          ", n:", total))) + 
  geom_boxplot(outlier.shape = NA, coef=0 ) +
  geom_jitter(width=0.1, size=2) +
  scale_color_manual(values=c("#ffbf00", "#66b032","#1982c4")) +
  theme_bw() +
scale_shape_manual(values=c(1,19)) +
  ggtitle('F2 offspring of F1 male (1/1) x female (0/0)') +
    facet_wrap( ~gt)  +  guides(color = "none") +
 theme(plot.margin = margin(1, 1, 1, 1, "cm"))

# warp by sex
p = df %>% dplyr::filter(total>10) %>%
  ggplot(aes(x = gt, y = freq, colour = gt, shape = adult_sisters, 
              text = paste("sample:", sample,
                          ", n:", total))) + 
  geom_boxplot(outlier.shape = NAf, coef=0 ) +
  geom_jitter(width=0.1, size=2) +
  scale_color_manual(values=c("#ffbf00", "#66b032","#1982c4")) +
  theme_bw() +
scale_shape_manual(values=c(1,19)) +
  ggtitle('F2 offspring of F1 male (1/1) x female (0/0)') +
    facet_wrap( ~sex)  +  guides(color = "none") +
 theme(plot.margin = margin(1, 1, 1, 1, "cm"))
```

```{r,  echo=FALSE,  out.width="100%", out.height="50%"}
ggplotly(p) %>% 
  layout(legend = list(orientation = "h", x = 0, y = -0.1),
         title = list(text = paste0('F2 offspring of F1 male (1/1) x female (0/0), ',
                                    '<br>',
                                    paste0("site count= ", sum(df$n)))))
```
only family 63 has sites in F2 for this cross.   
something is wrong...




# Informative crosses:
analyse "normal" and "abnormal" F2 males separately

## Crossing homozygotic male, with heterozygotic female
### (5) F1 male (0/0) x female (0/1)
```{r}
# Prepare table with observed and expected counts, per family:
# define a list to put all the data frames in
obs <- list()

for (fam in family) {
 
obs[[fam]] <- table %>%
  dplyr::select(starts_with(fam)) %>%
        dplyr::filter_at(vars(matches("_fnd")), all_vars(. == "0/0")) %>% # force F0 female to be homo, like her son
dplyr::filter_at(vars(matches("_son")), all_vars(. == "0/0")) %>%
  dplyr::filter_at(vars(matches("_dat")), all_vars(. == "0/1")) %>% 
  dplyr::select(contains("grn")) %>%
  tidyr::pivot_longer(everything())  %>% 
  dplyr::rename(sample = name, gt = value) %>%
  dplyr::count(sample, gt, .drop = FALSE) %>% 
  dplyr::filter(gt %in% c("0/0", "1/1", "0/1")) %>%
  mutate(n = as.numeric(n)) #%>%
#  group_by(sample,gt) %>%
 # mutate(total = as.numeric(sum(n)))
}

# bind all families together, to a final data frame containing all observed counts
observed <- do.call("rbind", obs) %>% mutate(sample = as.character(sample))
# replace "NA" with zero value when gt doesn't exists
samples <- data.frame(sample = rep(unique(observed$sample), each =3), gt = rep(c("0/0", "0/1", "1/1"))) # add all possible genotypes

df <- left_join(samples, observed, by=c("sample","gt")) %>%  
    replace(is.na(.),0) %>% # replace "NA" with zero value when gt doesn't exists 
    dplyr::group_by(sample) %>% 
    dplyr::mutate(total = as.numeric(sum(n))) %>% 
    dplyr::mutate(freq=round(n/total, digit=2)) %>%
    left_join(meta, by="sample") 

# order the stages 
df$sex <- factor(df$sex, level=c("female", "male", "Nymph", "Egg"))

# wrap by gt
p = df %>% dplyr::filter(total>10) %>%
  ggplot(aes(x = sex, y = freq, colour = gt, shape = adult_sisters, 
              text = paste("sample:", sample,
                          ", n:", total))) + 
  geom_boxplot(outlier.shape = NA, coef=0 ) +
  geom_jitter(width=0.1, size=2) +
  scale_color_manual(values=c("#ffbf00", "#66b032","#1982c4")) +
  theme_bw() +
scale_shape_manual(values=c(1,19)) +
  ggtitle('F2 offspring of F1 male (0/0) x female (0/1)') +
    facet_wrap( ~gt)  +  guides(color = "none") +
 theme(plot.margin = margin(1, 1, 1, 1, "cm"))

# warp by sex
p = df %>% dplyr::filter(total>10) %>%
  ggplot(aes(x = gt, y = freq, colour = gt, shape = adult_sisters, 
              text = paste("sample:", sample,
                          ", n:", total))) + 
  geom_boxplot(outlier.shape = NAf, coef=0 ) +
  geom_jitter(width=0.1, size=2) +
  scale_color_manual(values=c("#ffbf00", "#66b032","#1982c4")) +
  theme_bw() +
scale_shape_manual(values=c(1,19)) +
  ggtitle('F2 offspring of F1 male (0/0) x female (0/1)') +
    facet_wrap( ~sex)  +  guides(color = "none") +
 theme(plot.margin = margin(1, 1, 1, 1, "cm"))
```

#### Abnormal males
240_241c_grnson, 400_401a_grnson,412_413a_grnson, 426_427b_grnson, 458_459a_grnson, 46_47d_grnson

```{r,  echo=FALSE,  out.width="100%", out.height="50%"}
ggplotly(p) %>% 
  layout(legend = list(orientation = "h", x = 0, y = -0.1),
         title = list(text = paste0('F2 offspring of F1 male (0/0) x female (0/1), ',
                                    '<br>',
                                    paste0("site count= ", sum(df$n)))))
```

### (6) F1 male (1/1) x female (0/1)
```{r}
# Prepare table with observed and expected counts, per family:
# define a list to put all the data frames in
obs <- list()

for (fam in family) {
 
obs[[fam]] <- table %>%
  dplyr::select(starts_with(fam)) %>%
        dplyr::filter_at(vars(matches("_fnd")), all_vars(. == "1/1")) %>% # force F0 female to be homo, like her son
dplyr::filter_at(vars(matches("_son")), all_vars(. == "1/1")) %>%
  dplyr::filter_at(vars(matches("_dat")), all_vars(. == "0/1")) %>% 
  dplyr::select(contains("grn")) %>%
  tidyr::pivot_longer(everything())  %>% 
  dplyr::rename(sample = name, gt = value) %>%
  dplyr::count(sample, gt, .drop = FALSE) %>% 
  dplyr::filter(gt %in% c("0/0", "1/1", "0/1")) %>%
  mutate(n = as.numeric(n)) #%>%
#  group_by(sample,gt) %>%
 # mutate(total = as.numeric(sum(n)))
}

# bind all families together, to a final data frame containing all observed counts
observed <- do.call("rbind", obs) %>% mutate(sample = as.character(sample))
# replace "NA" with zero value when gt doesn't exists
samples <- data.frame(sample = rep(unique(observed$sample), each =3), gt = rep(c("0/0", "0/1", "1/1"))) # add all possible genotypes

df <- left_join(samples, observed, by=c("sample","gt")) %>%  
    replace(is.na(.),0) %>% # replace "NA" with zero value when gt doesn't exists 
    dplyr::group_by(sample) %>% 
    dplyr::mutate(total = as.numeric(sum(n))) %>% 
    dplyr::mutate(freq=round(n/total, digit=2)) %>%
    left_join(meta, by="sample") 

# order the stages 
df$sex <- factor(df$sex, level=c("female", "male", "Nymph", "Egg"))

# wrap by gt
p = df %>% dplyr::filter(total>10) %>%
  ggplot(aes(x = sex, y = freq, colour = gt, shape = adult_sisters, 
              text = paste("sample:", sample,
                          ", n:", total))) + 
  geom_boxplot(outlier.shape = NA, coef=0 ) +
  geom_jitter(width=0.1, size=2) +
  scale_color_manual(values=c("#ffbf00", "#66b032","#1982c4")) +
  theme_bw() +
scale_shape_manual(values=c(1,19)) +
  ggtitle('F2 offspring of F1 male (1/1) x female (0/1)') +
    facet_wrap( ~gt)  +  guides(color = "none") +
 theme(plot.margin = margin(1, 1, 1, 1, "cm"))

# warp by sex
p = df %>% dplyr::filter(total>10) %>%
  ggplot(aes(x = gt, y = freq, colour = gt, shape = adult_sisters, 
              text = paste("sample:", sample,
                          ", n:", total))) + 
  geom_boxplot(outlier.shape = NAf, coef=0 ) +
  geom_jitter(width=0.1, size=2) +
  scale_color_manual(values=c("#ffbf00", "#66b032","#1982c4")) +
  theme_bw() +
scale_shape_manual(values=c(1,19)) +
  ggtitle('F2 offspring of F1 male (1/1) x female (0/1)') +
    facet_wrap( ~sex)  +  guides(color = "none") +
 theme(plot.margin = margin(1, 1, 1, 1, "cm"))
```

```{r,  echo=FALSE,  out.width="100%", out.height="50%"}

ggplotly(p) %>% 
  layout(legend = list(orientation = "h", x = 0, y = -0.1),
         title = list(text = paste0('F2 offspring of F1 male (1/1) x female (0/1), ',
                                    '<br>',
                                    paste0("site count= ", sum(df$count)))))
```

## Crossing heterozygotic male, with homozygotic female
### (7) F1 male (0/1) x female (0/0)
```{r}
# Prepare table with observed and expected counts, per family:
# define a list to put all the data frames in
obs <- list()

for (fam in family) {
 
obs[[fam]] <- table %>%
  dplyr::select(starts_with(fam)) %>%
        dplyr::filter_at(vars(matches("_fnd")), all_vars(. == "0/1")) %>% # force F0 female to be homo, like her son
dplyr::filter_at(vars(matches("_son")), all_vars(. == "0/1")) %>%
  dplyr::filter_at(vars(matches("_dat")), all_vars(. == "0/0")) %>% 
  dplyr::select(contains("grn")) %>%
  tidyr::pivot_longer(everything())  %>% 
  dplyr::rename(sample = name, gt = value) %>%
  dplyr::count(sample, gt, .drop = FALSE) %>% 
  dplyr::filter(gt %in% c("0/0", "1/1", "0/1")) %>%
  mutate(n = as.numeric(n)) #%>%
#  group_by(sample,gt) %>%
 # mutate(total = as.numeric(sum(n)))
}

# bind all families together, to a final data frame containing all observed counts
observed <- do.call("rbind", obs) %>% mutate(sample = as.character(sample))
# replace "NA" with zero value when gt doesn't exists
samples <- data.frame(sample = rep(unique(observed$sample), each =3), gt = rep(c("0/0", "0/1", "1/1"))) # add all possible genotypes

df <- left_join(samples, observed, by=c("sample","gt")) %>%  
    replace(is.na(.),0) %>% # replace "NA" with zero value when gt doesn't exists 
    dplyr::group_by(sample) %>% 
    dplyr::mutate(total = as.numeric(sum(n))) %>% 
    dplyr::mutate(freq=round(n/total, digit=2)) %>%
    left_join(meta, by="sample") 

# order the stages 
df$sex <- factor(df$sex, level=c("female", "male", "Nymph", "Egg"))

# wrap by gt
p = df %>% dplyr::filter(total>10) %>%
  ggplot(aes(x = sex, y = freq, colour = gt, shape = adult_sisters, 
              text = paste("sample:", sample,
                          ", n:", total))) + 
  geom_boxplot(outlier.shape = NA, coef=0 ) +
  geom_jitter(width=0.1, size=2) +
  scale_color_manual(values=c("#ffbf00", "#66b032","#1982c4")) +
  theme_bw() +
scale_shape_manual(values=c(1,19)) +
  ggtitle('F2 offspring of F1 male (0/1) x female (0/10') +
    facet_wrap( ~gt)  +  guides(color = "none") +
 theme(plot.margin = margin(1, 1, 1, 1, "cm"))

# warp by sex
p = df %>% dplyr::filter(total>10) %>%
  ggplot(aes(x = gt, y = freq, colour = gt, shape = adult_sisters, 
              text = paste("sample:", sample,
                          ", n:", total))) + 
  geom_boxplot(outlier.shape = NAf, coef=0 ) +
  geom_jitter(width=0.1, size=2) +
  scale_color_manual(values=c("#ffbf00", "#66b032","#1982c4")) +
  theme_bw() +
scale_shape_manual(values=c(1,19)) +
  ggtitle('F2 offspring of F1 male (0/1) x female (0/0)') +
    facet_wrap( ~sex)  +  guides(color = "none") +
 theme(plot.margin = margin(1, 1, 1, 1, "cm"))
```

```{r,  echo=FALSE,  out.width="100%", out.height="50%"}
ggplotly(p) %>% 
  layout(legend = list(orientation = "h", x = 0, y = -0.1),
         title = list(text = paste0('F2 offspring of F1 male (0/1) x female (0/0), ',
                                    '<br>',
                                    paste0("site count= ", sum(df$count)))))
```

### (8) F1 male (0/1) x female (1/1)
```{r}
# Prepare table with observed and expected counts, per family:
# define a list to put all the data frames in
obs <- list()

for (fam in family) {
 
obs[[fam]] <- table %>%
  dplyr::select(starts_with(fam)) %>%
        dplyr::filter_at(vars(matches("_fnd")), all_vars(. == "0/1")) %>% # force F0 female to be homo, like her son
dplyr::filter_at(vars(matches("_son")), all_vars(. == "0/1")) %>%
  dplyr::filter_at(vars(matches("_dat")), all_vars(. == "1/1")) %>% 
  dplyr::select(contains("grn")) %>%
  tidyr::pivot_longer(everything())  %>% 
  dplyr::rename(sample = name, gt = value) %>%
  dplyr::count(sample, gt, .drop = FALSE) %>% 
  dplyr::filter(gt %in% c("0/0", "1/1", "0/1")) %>%
  mutate(n = as.numeric(n)) #%>%
#  group_by(sample,gt) %>%
 # mutate(total = as.numeric(sum(n)))
}

# bind all families together, to a final data frame containing all observed counts
observed <- do.call("rbind", obs) %>% mutate(sample = as.character(sample))
# replace "NA" with zero value when gt doesn't exists
samples <- data.frame(sample = rep(unique(observed$sample), each =3), gt = rep(c("0/0", "0/1", "1/1"))) # add all possible genotypes

df <- left_join(samples, observed, by=c("sample","gt")) %>%  
    replace(is.na(.),0) %>% # replace "NA" with zero value when gt doesn't exists 
    dplyr::group_by(sample) %>% 
    dplyr::mutate(total = as.numeric(sum(n))) %>% 
    dplyr::mutate(freq=round(n/total, digit=2)) %>%
    left_join(meta, by="sample") 

# order the stages 
df$sex <- factor(df$sex, level=c("female", "male", "Nymph", "Egg"))

# wrap by gt
p = df %>% dplyr::filter(total>10) %>%
  ggplot(aes(x = sex, y = freq, colour = gt, shape = adult_sisters, 
              text = paste("sample:", sample,
                          ", n:", total))) + 
  geom_boxplot(outlier.shape = NA, coef=0 ) +
  geom_jitter(width=0.1, size=2) +
  scale_color_manual(values=c("#ffbf00", "#66b032","#1982c4")) +
  theme_bw() +
scale_shape_manual(values=c(1,19)) +
  ggtitle('F2 offspring of F1 male (0/1) x female (1/1') +
    facet_wrap( ~gt)  +  guides(color = "none") +
 theme(plot.margin = margin(1, 1, 1, 1, "cm"))

# warp by sex
p = df %>% dplyr::filter(total>10) %>%
  ggplot(aes(x = gt, y = freq, colour = gt, shape = adult_sisters, 
              text = paste("sample:", sample,
                          ", n:", total))) + 
  geom_boxplot(outlier.shape = NAf, coef=0 ) +
  geom_jitter(width=0.1, size=2) +
  scale_color_manual(values=c("#ffbf00", "#66b032","#1982c4")) +
  theme_bw() +
scale_shape_manual(values=c(1,19)) +
  ggtitle('F2 offspring of F1 male (0/1) x female (1/1)') +
    facet_wrap( ~sex)  +  guides(color = "none") +
 theme(plot.margin = margin(1, 1, 1, 1, "cm"))
```

```{r,  echo=FALSE,  out.width="100%", out.height="50%"}
ggplotly(p) %>% 
  layout(legend = list(orientation = "h", x = 0, y = -0.1),
         title = list(text = paste0('F2 offspring of F1 male (0/1) x female (1/1), ',
                                    '<br>',
                                    paste0("site count= ", sum(df$count)))))
```

## Crossing heterozygotic male and female
### (9) F1 male (0/1) x female (0/1)
```{r}
# Prepare table with observed and expected counts, per family:
# define a list to put all the data frames in
obs <- list()

for (fam in family) {
 
obs[[fam]] <- table %>%
  dplyr::select(starts_with(fam)) %>%
        dplyr::filter_at(vars(matches("_fnd")), all_vars(. == "0/1")) %>% # force F0 female to be homo, like her son
dplyr::filter_at(vars(matches("_son")), all_vars(. == "0/1")) %>%
  dplyr::filter_at(vars(matches("_dat")), all_vars(. == "0/1")) %>% 
  dplyr::select(contains("grn")) %>%
  tidyr::pivot_longer(everything())  %>% 
  dplyr::rename(sample = name, gt = value) %>%
  dplyr::count(sample, gt, .drop = FALSE) %>% 
  dplyr::filter(gt %in% c("0/0", "1/1", "0/1")) %>%
  mutate(n = as.numeric(n)) #%>%
#  group_by(sample,gt) %>%
 # mutate(total = as.numeric(sum(n)))
}

# bind all families together, to a final data frame containing all observed counts
observed <- do.call("rbind", obs) %>% mutate(sample = as.character(sample))
# replace "NA" with zero value when gt doesn't exists
samples <- data.frame(sample = rep(unique(observed$sample), each =3), gt = rep(c("0/0", "0/1", "1/1"))) # add all possible genotypes

df <- left_join(samples, observed, by=c("sample","gt")) %>%  
    replace(is.na(.),0) %>% # replace "NA" with zero value when gt doesn't exists 
    dplyr::group_by(sample) %>% 
    dplyr::mutate(total = as.numeric(sum(n))) %>% 
    dplyr::mutate(freq=round(n/total, digit=2)) %>%
    left_join(meta, by="sample") 

# order the stages 
df$sex <- factor(df$sex, level=c("female", "male", "Nymph", "Egg"))

# wrap by gt
p = df %>% dplyr::filter(total>10) %>%
  ggplot(aes(x = sex, y = freq, colour = gt, shape = adult_sisters, 
              text = paste("sample:", sample,
                          ", n:", total))) + 
  geom_boxplot(outlier.shape = NA, coef=0 ) +
  geom_jitter(width=0.1, size=2) +
  scale_color_manual(values=c("#ffbf00", "#66b032","#1982c4")) +
  theme_bw() +
scale_shape_manual(values=c(1,19)) +
  ggtitle('F2 offspring of F1 male (0/1) x female (0/1') +
    facet_wrap( ~gt)  +  guides(color = "none") +
 theme(plot.margin = margin(1, 1, 1, 1, "cm"))

# warp by sex
p = df %>% dplyr::filter(total>10) %>%
  ggplot(aes(x = gt, y = freq, colour = gt, shape = adult_sisters, 
              text = paste("sample:", sample,
                          ", n:", total))) + 
  geom_boxplot(outlier.shape = NAf, coef=0 ) +
  geom_jitter(width=0.1, size=2) +
  scale_color_manual(values=c("#ffbf00", "#66b032","#1982c4")) +
  theme_bw() +
scale_shape_manual(values=c(1,19)) +
  ggtitle('F2 offspring of F1 male (0/1) x female (0/1)') +
    facet_wrap( ~sex)  +  guides(color = "none") +
 theme(plot.margin = margin(1, 1, 1, 1, "cm"))
```

```{r,  echo=FALSE,  out.width="100%", out.height="50%"}
ggplotly(p) %>% 
  layout(legend = list(orientation = "h", x = 0, y = -0.1),
         title = list(text = paste0('F2 offspring of F1 male (0/1) x female (0/1), ',
                                    '<br>',
                                    paste0("site count= ", sum(df$n)))))
```

