---
title: "pedigree-hetero-count-Q15000"
editor_options: 
  chunk_output_type: console
author: "Nurit Eliash"
date: "1/24/2022"
output:
  html_document:
    code_folding: hide
    theme: cerulean
    toc: yes
    toc_depth: 5
    toc_float:
      collapsed: yes
      df_print: paged
---

```{=html}
<style type="text/css">
  body{
  font-size: 12pt;
}
</style>
```

### load libraries
```{r libraries, echo=TRUE, message=FALSE, warning=FALSE}
library("tidyverse")
library("plyr")
library("dplyr")
library("ggplot2")
library("scales")
library("ggpubr")
library("gridExtra")
library("grid")
library("GGally")
library("vcfR") # for extracting genotype data from a vcf file
library("data.table")
library("stringr")
library("janitor")
library("gmodels")
library("rstatix")
library("freqtables")
library("cowplot")
library("patchwork") # for gathering the plots
library("pastecs") # for stats stat.desc()
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE,
                     fig.width = 10,
                      fig.asp = 0.9,
                      out.width = "100%")
#fig.width = 6,fig.asp = 0.8,out.width = "100%"
```

the VCF was filtered :  
```{r eval=FALSE}
vcftools --vcf snponly_freebayes.vcf --chr NW_019211454.1 --chr NW_019211455.1 --chr NW_019211456.1 --chr NW_019211457.1 --chr NW_019211458.1 --chr NW_019211459.1 --chr NW_019211460.1 --max-alleles 2 --minQ 15000 --minDP 16 --maxDP 40 --max-missing 0.5 --maf 0.2 --recode --recode-INFO-all --out Q15000BIALLDP16HDP40mis.5maf.2Chr7  
```

## Load Variant Call Format (VCF) file.

```{r}
vcf <- read.vcfR("/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/data/vcf_filter/minQ_filter/Q15000BIALLDP16HDP40mis.5maf.2Chr7.recode.vcf", verbose = FALSE )
vcf
# extract the genotype for each site in each individual
gt <- extract.gt(vcf, element = "GT") 
gt <- as.data.frame(t(gt)) %>%
  rownames_to_column("ID")
#clean the ID names
gt$ID <- sub("_[^_]+$", "", gt$ID)

table <-  gt %>% 
  t() %>%
  as.data.frame() %>%
  row_to_names(row_number = 1) %>% 
  dplyr::select(contains(c("son", "dat", "fnd"))) # keep only adults of F0, F1 and F2 

# set the families (include only families with at least one adult F2)
family = grep("grndat|grnson",gt$ID, value=TRUE) %>%
  str_extract("[^_]+")  %>%
  unique()

# or, include all F2 samples, but indicate if they have an adult sister (may indicate if the F1 female was fertilized)
#family = grep("grn",gt$ID, value=TRUE) %>%
#  str_extract("[^_]+")  %>%
#  unique()
```

------------------------------------------------------------------------

# Crossing homozygotic male, with heterozygotic female

## (5) F1 male (0/0) x female (0/1)

### step-by-step filtering:

#### When do we lose most of the sites in the F2 mites?

We saw theres much difference in the number of sites in each cross.  
to check this issue,we checked which of the filters is reducing most of the sites

#### F2 genotypes, no filters
```{r}
# define a list to put all the data frames in
obs <- list()

for (fam in family) {
 
obs[[fam]] <- table %>%
  dplyr::select(starts_with(fam)) %>%
  #  dplyr::filter_at(vars(matches("_fnd")), all_vars(. == "0/0")) %>% # force F0 female to be homo, like her son
 # dplyr::filter_at(vars(matches("_son")), all_vars(. == "0/0")) %>%
  #dplyr::filter_at(vars(matches("_dat")), all_vars(. == "0/1")) %>% 
  dplyr::select(contains("grn")) %>%
  tidyr::pivot_longer(everything())  %>% 
  dplyr::rename(sample = name, gt = value) %>%
  dplyr::count(sample, gt, .drop = FALSE) %>% 
  dplyr::filter(gt %in% c("0/0", "1/1", "0/1")) %>%
  mutate(n = as.numeric(n)) %>%
  group_by(sample) %>%
  mutate(total = as.numeric(sum(n))) %>%
  mutate(sex = case_when(
    grepl("son", sample) ~ "male",
    grepl("dat", sample) ~ "female"))
}

# bind all families together, to a final data frame containing all observed counts
observed <- do.call("rbind", obs) %>% mutate(sample = as.character(sample))
samples <- data.frame(sample = rep(unique(observed$sample), each =3), gt = rep(c("0/0", "0/1", "1/1")))

samples_obs <- left_join(samples, observed, by=c("sample","gt")) %>% mutate(sex = case_when(
    grepl("son", sample) ~ "male",
    grepl("dat", sample) ~ "female")) %>%
    group_by(sample) %>%
    replace(is.na(.), 0) %>%
    mutate(total = as.numeric(sum(n))) %>%
    dplyr::mutate(prop = n/total) 
  
pooled_obs_count =  samples_obs %>% 
    dplyr::filter(total>=10) %>%
    filter(gt =="0/0") %>%
    group_by(sex) %>% 
    mutate(sites = sum(total)) %>% 
   mutate(count_sample = n_distinct(sample)) %>%
   select(c(sex,sites,count_sample)) %>% 
    unique()

pooled_obs_count

p_all_f2_bar = samples_obs %>% dplyr::filter(total>=10) %>%
    ggplot(aes(fill=gt, y=prop, x=sex)) + 
    geom_bar(position="fill", stat="identity", ) +
    labs(fill = "Genotype") + 
    theme_classic() +
    theme(axis.text.x = element_text(size=12),
        axis.text.y = element_text(size=12)) +
     scale_fill_manual(values=c("#ffbf00", "#66b032","#1982c4"))+
    # theme(legend.position = "none")+
  ggtitle("No filter",          
          subtitle = paste("F2 female = ", filter(pooled_obs_count, sex == 'female')$sites, ", male =", filter(pooled_obs_count, sex == 'male')$sites))
```

#### F0 female, (0/0)
```{r}
# define a list to put all the data frames in
obs <- list()

for (fam in family) {
 
obs[[fam]] <- table %>%
  dplyr::select(starts_with(fam)) %>%
    dplyr::filter_at(vars(matches("_fnd")), all_vars(. == "0/0")) %>% # force F0 female to be homo, like her son
 # dplyr::filter_at(vars(matches("_son")), all_vars(. == "0/0")) %>%
  #dplyr::filter_at(vars(matches("_dat")), all_vars(. == "0/1")) %>% 
  dplyr::select(contains("grn")) %>%
  tidyr::pivot_longer(everything())  %>% 
  dplyr::rename(sample = name, gt = value) %>%
  dplyr::count(sample, gt, .drop = FALSE) %>% 
  dplyr::filter(gt %in% c("0/0", "1/1", "0/1")) %>%
  mutate(n = as.numeric(n)) %>%
  group_by(sample) %>%
  mutate(total = as.numeric(sum(n))) %>%
  mutate(sex = case_when(
    grepl("son", sample) ~ "male",
    grepl("dat", sample) ~ "female"))
}

# bind all families together, to a final data frame containing all observed counts
observed <- do.call("rbind", obs) %>% mutate(sample = as.character(sample))
samples <- data.frame(sample = rep(unique(observed$sample), each =3), gt = rep(c("0/0", "0/1", "1/1")))

samples_obs <- left_join(samples, observed, by=c("sample","gt")) %>% mutate(sex = case_when(
    grepl("son", sample) ~ "male",
    grepl("dat", sample) ~ "female")) %>%
    group_by(sample) %>%
    replace(is.na(.), 0) %>%
    mutate(total = as.numeric(sum(n))) %>%
    dplyr::mutate(prop = n/total) 
  
pooled_obs_count =  samples_obs %>% 
    dplyr::filter(total>=10) %>%
    filter(gt =="0/0") %>%
    group_by(sex) %>% 
    mutate(sites = sum(total)) %>% 
   mutate(count_sample = n_distinct(sample)) %>%
   select(c(sex,sites,count_sample)) %>% 
    unique()

pooled_obs_count

p_F0_bar = samples_obs %>% dplyr::filter(total>=10) %>%
    ggplot(aes(fill=gt, y=prop, x=sex)) + 
    geom_bar(position="fill", stat="identity", ) +
    labs(fill = "Genotype") + 
    theme_classic() +
    theme(axis.text.x = element_text(size=12),
        axis.text.y = element_text(size=12)) +
     scale_fill_manual(values=c("#ffbf00", "#66b032","#1982c4"))+
     theme(legend.position = "none")+
  ggtitle("F0 female (0/0)",          
          subtitle = paste("F2 female = ", filter(pooled_obs_count, sex == 'female')$sites, ", male =", filter(pooled_obs_count, sex == 'male')$sites))
```

#### F0 female, (0/0); F1 male (0/0)
```{r}
# define a list to put all the data frames in
obs <- list()

for (fam in family) {
 
obs[[fam]] <- table %>%
  dplyr::select(starts_with(fam)) %>%
    dplyr::filter_at(vars(matches("_fnd")), all_vars(. == "0/0")) %>% # force F0 female to be homo, like her son
  dplyr::filter_at(vars(matches("_son")), all_vars(. == "0/0")) %>%
  #dplyr::filter_at(vars(matches("_dat")), all_vars(. == "0/1")) %>% 
  dplyr::select(contains("grn")) %>%
  tidyr::pivot_longer(everything())  %>% 
  dplyr::rename(sample = name, gt = value) %>%
  dplyr::count(sample, gt, .drop = FALSE) %>% 
  dplyr::filter(gt %in% c("0/0", "1/1", "0/1")) %>%
  mutate(n = as.numeric(n)) %>%
  group_by(sample) %>%
  mutate(total = as.numeric(sum(n))) %>%
  mutate(sex = case_when(
    grepl("son", sample) ~ "male",
    grepl("dat", sample) ~ "female"))
}

# bind all families together, to a final data frame containing all observed counts
observed <- do.call("rbind", obs) %>% mutate(sample = as.character(sample))
samples <- data.frame(sample = rep(unique(observed$sample), each =3), gt = rep(c("0/0", "0/1", "1/1")))

samples_obs <- left_join(samples, observed, by=c("sample","gt")) %>% mutate(sex = case_when(
    grepl("son", sample) ~ "male",
    grepl("dat", sample) ~ "female")) %>%
    group_by(sample) %>%
    replace(is.na(.), 0) %>%
    mutate(total = as.numeric(sum(n))) %>%
    dplyr::mutate(prop = n/total) 
  
pooled_obs_count =  samples_obs %>% 
    dplyr::filter(total>=10) %>%
    filter(gt =="0/0") %>%
    group_by(sex) %>% 
    mutate(sites = sum(total)) %>% 
   mutate(count_sample = n_distinct(sample)) %>%
   select(c(sex,sites,count_sample)) %>% 
    unique()

pooled_obs_count

p_F0_F1_male_bar = samples_obs %>% dplyr::filter(total>=10) %>%
    ggplot(aes(fill=gt, y=prop, x=sex)) + 
    geom_bar(position="fill", stat="identity", ) +
    labs(fill = "Genotype") + 
    theme_classic() +
    theme(axis.text.x = element_text(size=12),
        axis.text.y = element_text(size=12)) +
     scale_fill_manual(values=c("#ffbf00", "#66b032","#1982c4"))+
     theme(legend.position = "none")+
  ggtitle("F0 female (0/0), F1 male (0/0)",          
          subtitle = paste("F2 female = ", filter(pooled_obs_count, sex == 'female')$sites, ", male =", filter(pooled_obs_count, sex == 'male')$sites))
```

#### F0 female, (0/0); F1 female (0/1)
```{r}
# define a list to put all the data frames in
obs <- list()

for (fam in family) {
 
obs[[fam]] <- table %>%
  dplyr::select(starts_with(fam)) %>%
    dplyr::filter_at(vars(matches("_fnd")), all_vars(. == "0/0")) %>% # force F0 female to be homo, like her son
  #dplyr::filter_at(vars(matches("_son")), all_vars(. == "0/0")) %>%
  dplyr::filter_at(vars(matches("_dat")), all_vars(. == "0/1")) %>% 
  dplyr::select(contains("grn")) %>%
  tidyr::pivot_longer(everything())  %>% 
  dplyr::rename(sample = name, gt = value) %>%
  dplyr::count(sample, gt, .drop = FALSE) %>% 
  dplyr::filter(gt %in% c("0/0", "1/1", "0/1")) %>%
  mutate(n = as.numeric(n)) %>%
  group_by(sample) %>%
  mutate(total = as.numeric(sum(n))) %>%
  mutate(sex = case_when(
    grepl("son", sample) ~ "male",
    grepl("dat", sample) ~ "female"))
}

# bind all families together, to a final data frame containing all observed counts
observed <- do.call("rbind", obs) %>% mutate(sample = as.character(sample))
samples <- data.frame(sample = rep(unique(observed$sample), each =3), gt = rep(c("0/0", "0/1", "1/1")))

samples_obs <- left_join(samples, observed, by=c("sample","gt")) %>% mutate(sex = case_when(
    grepl("son", sample) ~ "male",
    grepl("dat", sample) ~ "female")) %>%
    group_by(sample) %>%
    replace(is.na(.), 0) %>%
    mutate(total = as.numeric(sum(n))) %>%
    dplyr::mutate(prop = n/total) 

pooled_obs_count =  samples_obs %>% 
    dplyr::filter(total>=10) %>%
    filter(gt =="0/0") %>%
    group_by(sex) %>% 
    mutate(sites = sum(total)) %>% 
   mutate(count_sample = n_distinct(sample)) %>%
   select(c(sex,sites,count_sample)) %>% 
    unique()

pooled_obs_count

p_F0_F1_fem_bar = samples_obs %>% dplyr::filter(total>=10) %>%
    ggplot(aes(fill=gt, y=prop, x=sex)) + 
    geom_bar(position="fill", stat="identity", ) +
    labs(fill = "Genotype") + 
    theme_classic() +
    theme(axis.text.x = element_text(size=12),
        axis.text.y = element_text(size=12)) +
     scale_fill_manual(values=c("#ffbf00", "#66b032","#1982c4"))+
     theme(legend.position = "none")+
  ggtitle("F0 female (0/0), F1 female (0/1)",          
          subtitle = paste("F2 female = ", filter(pooled_obs_count, sex == 'female')$sites, ", male =", filter(pooled_obs_count, sex == 'male')$sites))
```

#### F0 female, (0/0); F1 male (0/0); F1 female (0/1)
```{r}
# define a list to put all the data frames in
obs <- list()

for (fam in family) {
 
obs[[fam]] <- table %>%
  dplyr::select(starts_with(fam)) %>%
    dplyr::filter_at(vars(matches("_fnd")), all_vars(. == "0/0")) %>% # force F0 female to be homo, like her son
  dplyr::filter_at(vars(matches("_son")), all_vars(. == "0/0")) %>%
  dplyr::filter_at(vars(matches("_dat")), all_vars(. == "0/1")) %>% 
  dplyr::select(contains("grn")) %>%
  tidyr::pivot_longer(everything())  %>% 
  dplyr::rename(sample = name, gt = value) %>%
  dplyr::count(sample, gt, .drop = FALSE) %>% 
  dplyr::filter(gt %in% c("0/0", "1/1", "0/1")) %>%
  mutate(n = as.numeric(n)) %>%
  group_by(sample) %>%
  mutate(total = as.numeric(sum(n))) %>%
  mutate(sex = case_when(
    grepl("son", sample) ~ "male",
    grepl("dat", sample) ~ "female"))
}

# bind all families together, to a final data frame containing all observed counts
observed <- do.call("rbind", obs) %>% mutate(sample = as.character(sample))
samples <- data.frame(sample = rep(unique(observed$sample), each =3), gt = rep(c("0/0", "0/1", "1/1")))

samples_obs_00_01 <- left_join(samples, observed, by=c("sample","gt")) %>% 
  mutate(sex = case_when(
    grepl("son", sample) ~ "male",
    grepl("dat", sample) ~ "female")) %>%
    group_by(sample) %>%
    replace(is.na(.), 0) %>%
    mutate(total = as.numeric(sum(n))) %>%
    dplyr::mutate(prop = n/total) %>%
  mutate(parent = case_when(
    grepl("0/1", gt) ~ "maternal",
    grepl("0/0", gt) ~ "paternal",
    grepl("1/1", gt) ~ "mistake"))

pooled_obs_count =  samples_obs_00_01 %>% 
    dplyr::filter(total>=10) %>%
    filter(gt =="0/0") %>%
    group_by(sex) %>% 
    mutate(sites = sum(total)) %>% 
   mutate(count_sample = n_distinct(sample)) %>%
   select(c(sex,sites,count_sample)) %>% 
    unique()

pooled_obs_count

p_00_01_bar = samples_obs_00_01 %>% dplyr::filter(total>=10) %>%
    ggplot(aes(fill=gt, y=prop, x=sex)) + 
    geom_bar(position="fill", stat="identity", ) +
    labs(fill = "Genotype") + 
    theme_classic() +
    theme(axis.text.x = element_text(size=12),
        axis.text.y = element_text(size=12)) +
     scale_fill_manual(values=c("#ffbf00", "#66b032","#1982c4"))+
     theme(legend.position = "none")+
  ggtitle("F0 female (0/0), F1 male (0/0), F1 female (0/1)",          
          subtitle = paste("F2 female = ", filter(pooled_obs_count, sex == 'female')$sites, ", male =", filter(pooled_obs_count, sex == 'male')$sites))

#density plot:
#samples_obs %>% 
 # dplyr::filter(total>=10) %>%  
 # dplyr::filter(sex == "male") %>%
 # dplyr::filter(gt == "0/1") %>% 
  #ggplot(aes(prop, fill = sex)) + geom_density(colour = "black", alpha = 0.5) + theme_light() +
  #ggtitle("F2 maternal genotype proportion",
   #       subtitle = "F0 female (0/0); F1 male (0/0); F1 female (0/1)")

# histogram plot (count)
p_00_01_hist = samples_obs_00_01 %>% 
  dplyr::filter(total>=10) %>%  
  dplyr::filter(gt == "0/1") %>%
  ggplot(aes(prop, fill = sex)) +
    geom_histogram(binwidth=.05, alpha=.7, position="identity") + theme_light() +
  ggtitle("F1 cross: male (0/0) x female (0/1)",
          subtitle = paste("F2 female = ", filter(pooled_obs_count, sex == 'female')$count_sample, ", male =", filter(pooled_obs_count, sex == 'male')$count_sample))

#write_csv(samples_obs_00_01, "/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/data/samples_prop_00_01.csv")
```

step-by-step filtering shows that the filtering of the F1 as 0/1 reduces most of the sites.  
This makes sense, and in corresponding with the varroa pedigree structure:  

- when F0 female is 0/0, for her daughter to be 0/1, the F0 male should be either 0/1, or 1/1
Both genotypes are mote rare then 0/0.  
therefore, when we ask for: F0 = 0/0, and F1 female = 0/1, we lose most of the sites.

This explains why we get such a high difference in the number of sites in F2 between the different crosses.  
```{r}
legend <- get_legend(p_all_f2_bar)  

# here the plots in a grid
prow <- plot_grid(p_all_f2_bar + theme(legend.position="none"),
           p_F0_bar + theme(legend.position="none")+ scale_y_continuous(),
           p_F0_F1_male_bar + theme(legend.position="none")+ scale_y_continuous(),
          p_F0_F1_fem_bar + theme(legend.position="none")+ scale_y_continuous(),
            p_00_01_bar + theme(legend.position="none")+ scale_y_continuous(),
          align = 'v',
           hjust = 1,
            vjust = 1.5,
           nrow = 3)

# here you add the legend
plot_grid( prow, legend, rel_widths = c(1, .2))
```

#### F0 female, (1/1); F1 male (1/1); F1 female (0/1)
```{r}
# define a list to put all the data frames in
obs <- list()

for (fam in family) {
 
obs[[fam]] <- table %>%
  dplyr::select(starts_with(fam)) %>%
    dplyr::filter_at(vars(matches("_fnd")), all_vars(. == "1/1")) %>% # force F0 female to be homo, like her son
  dplyr::filter_at(vars(matches("_son")), all_vars(. == "1/1")) %>%
  dplyr::filter_at(vars(matches("_dat")), all_vars(. == "0/1")) %>% 
  dplyr::select(contains("grn")) %>%
  tidyr::pivot_longer(everything())  %>% 
  dplyr::rename(sample = name, gt = value) %>%
  dplyr::count(sample, gt, .drop = FALSE) %>% 
  dplyr::filter(gt %in% c("0/0", "1/1", "0/1")) %>%
  mutate(n = as.numeric(n)) %>%
  group_by(sample) %>%
  mutate(total = as.numeric(sum(n))) %>%
  mutate(sex = case_when(
    grepl("son", sample) ~ "male",
    grepl("dat", sample) ~ "female"))
}

# bind all families together, to a final data frame containing all observed counts
observed <- do.call("rbind", obs) %>% mutate(sample = as.character(sample))
samples <- data.frame(sample = rep(unique(observed$sample), each =3), gt = rep(c("0/0", "0/1", "1/1")))

samples_obs_11_01 <- left_join(samples, observed, by=c("sample","gt")) %>% mutate(sex = case_when(
    grepl("son", sample) ~ "male",
    grepl("dat", sample) ~ "female")) %>%
    group_by(sample) %>%
    replace(is.na(.), 0) %>%
    mutate(total = as.numeric(sum(n))) %>%
    dplyr::mutate(prop = n/total) %>%
  mutate(parent = case_when(
    grepl("0/1", gt) ~ "maternal",
    grepl("1/1", gt) ~ "paternal",
    grepl("0/0", gt) ~ "mistake"))
  
pooled_obs_count =  samples_obs_11_01 %>% 
    dplyr::filter(total>=10) %>%
    filter(gt =="0/0") %>%
    group_by(sex) %>% 
    mutate(sites = sum(total)) %>% 
   mutate(count_sample = n_distinct(sample)) %>%
   select(c(sex,sites,count_sample)) %>% 
    unique()

pooled_obs_count

p_11_01_bar = samples_obs_11_01 %>% dplyr::filter(total>=10) %>%
    ggplot(aes(fill=gt, y=prop, x=sex)) + 
    geom_bar(position="fill", stat="identity", ) +
    labs(fill = "Genotype") + 
    theme_classic() +
    theme(axis.text.x = element_text(size=12),
        axis.text.y = element_text(size=12)) +
     scale_fill_manual(values=c("#ffbf00", "#66b032","#1982c4"))+
     theme(legend.position = "none")+
  ggtitle("F0 female, (1/1); F1 male (1/1); F1 female (0/1)")

# histogram plot (count)
p_11_01_hist = samples_obs_11_01 %>% 
  dplyr::filter(total>=10) %>%  
  dplyr::filter(gt == "0/1") %>%
  ggplot(aes(prop, fill = sex)) +
    geom_histogram(binwidth=.05, alpha=.7, position="identity") + theme_light() +
  ggtitle("F1 cross: male (1/1) x female (0/1)",
          subtitle = paste("F2 female = ", filter(pooled_obs_count, sex == 'female')$count_sample, ", male =", filter(pooled_obs_count, sex == 'male')$count_sample))

# save as csv to send to sasha
#write_csv(samples_obs_11_01, "/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/data/samples_prop_11_01.csv")
```

#### F0 female, (0/1); F1 male (0/1); F1 female (0/0)
```{r}
# define a list to put all the data frames in
obs <- list()

for (fam in family) {
 
obs[[fam]] <- table %>%
  dplyr::select(starts_with(fam)) %>%
    dplyr::filter_at(vars(matches("_fnd")), all_vars(. == "0/1")) %>% # force F0 female to be homo, like her son
  dplyr::filter_at(vars(matches("_son")), all_vars(. == "0/1")) %>%
  dplyr::filter_at(vars(matches("_dat")), all_vars(. == "0/0")) %>% 
  dplyr::select(contains("grn")) %>%
  tidyr::pivot_longer(everything())  %>% 
  dplyr::rename(sample = name, gt = value) %>%
  dplyr::count(sample, gt, .drop = FALSE) %>% 
  dplyr::filter(gt %in% c("0/0", "1/1", "0/1")) %>%
  mutate(n = as.numeric(n)) %>%
  group_by(sample) %>%
  mutate(total = as.numeric(sum(n))) %>%
  mutate(sex = case_when(
    grepl("son", sample) ~ "male",
    grepl("dat", sample) ~ "female"))
}

# bind all families together, to a final data frame containing all observed counts
observed <- do.call("rbind", obs) %>% mutate(sample = as.character(sample))
samples <- data.frame(sample = rep(unique(observed$sample), each =3), gt = rep(c("0/0", "0/1", "1/1")))

samples_obs_01_00 <- left_join(samples, observed, by=c("sample","gt")) %>% mutate(sex = case_when(
    grepl("son", sample) ~ "male",
    grepl("dat", sample) ~ "female")) %>%
    group_by(sample) %>%
    replace(is.na(.), 0) %>%
    mutate(total = as.numeric(sum(n))) %>%
    dplyr::mutate(prop = n/total) %>%
  mutate(parent = case_when(
    grepl("0/0", gt) ~ "maternal",
    grepl("0/1", gt) ~ "paternal",
    grepl("1/1", gt) ~ "mistake"))
  
pooled_obs_count =  samples_obs_01_00 %>% 
    dplyr::filter(total>=10) %>%
    filter(gt =="0/0") %>%
    group_by(sex) %>% 
    mutate(sites = sum(total)) %>% 
   mutate(count_sample = n_distinct(sample)) %>%
   select(c(sex,sites,count_sample)) %>% 
    unique()

pooled_obs_count

p_01_00_bar = samples_obs_01_00 %>% dplyr::filter(total>=10) %>%
    ggplot(aes(fill=gt, y=prop, x=sex)) + 
    geom_bar(position="fill", stat="identity", ) +
    labs(fill = "Genotype") + 
    theme_classic() +
    theme(axis.text.x = element_text(size=12),
        axis.text.y = element_text(size=12)) +
     scale_fill_manual(values=c("#ffbf00", "#66b032","#1982c4"))+
     theme(legend.position = "none")+
  ggtitle("F0 female, (0/1); F1 male (0/1); F1 female (0/0)")

# histogram plot (count)
p_01_00_hist = samples_obs_01_00 %>% 
  dplyr::filter(total>=10) %>%  
  dplyr::filter(gt == "0/0") %>%
  ggplot(aes(prop, fill = sex)) +
    geom_histogram(binwidth=.05, alpha=.7, position="identity") + theme_light() +
  ggtitle("F1 cross: male (0/1) x female (0/0)",
          subtitle = paste("F2 female = ", filter(pooled_obs_count, sex == 'female')$count_sample, ", male =", filter(pooled_obs_count, sex == 'male')$count_sample))

# save as csv to send to sasha
#write_csv(samples_obs_01_00, "/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/data/samples_prop_01_00.csv")
```

#### F0 female, (0/1); F1 male (0/1); F1 female (1/1)
```{r}
# define a list to put all the data frames in
obs <- list()

for (fam in family) {
 
obs[[fam]] <- table %>%
  dplyr::select(starts_with(fam)) %>%
    dplyr::filter_at(vars(matches("_fnd")), all_vars(. == "0/1")) %>% # force F0 female to be homo, like her son
  dplyr::filter_at(vars(matches("_son")), all_vars(. == "0/1")) %>%
  dplyr::filter_at(vars(matches("_dat")), all_vars(. == "1/1")) %>% 
  dplyr::select(contains("grn")) %>%
  tidyr::pivot_longer(everything())  %>% 
  dplyr::rename(sample = name, gt = value) %>%
  dplyr::count(sample, gt, .drop = FALSE) %>% 
  dplyr::filter(gt %in% c("0/0", "1/1", "0/1")) %>%
  mutate(n = as.numeric(n)) %>%
  group_by(sample) %>%
  mutate(total = as.numeric(sum(n))) %>%
  mutate(sex = case_when(
    grepl("son", sample) ~ "male",
    grepl("dat", sample) ~ "female"))
}

# bind all families together, to a final data frame containing all observed counts
observed <- do.call("rbind", obs) %>% mutate(sample = as.character(sample))
samples <- data.frame(sample = rep(unique(observed$sample), each =3), gt = rep(c("0/0", "0/1", "1/1")))

samples_obs_01_11 <- left_join(samples, observed, by=c("sample","gt")) %>% mutate(sex = case_when(
    grepl("son", sample) ~ "male",
    grepl("dat", sample) ~ "female")) %>%
    group_by(sample) %>%
    replace(is.na(.), 0) %>%
    mutate(total = as.numeric(sum(n))) %>%
    dplyr::mutate(prop = n/total) %>%
  mutate(parent = case_when(
    grepl("1/1", gt) ~ "maternal",
    grepl("0/1", gt) ~ "paternal",
    grepl("0/0", gt) ~ "mistake"))
  
pooled_obs_count =  samples_obs_01_11 %>% 
    dplyr::filter(total>=10) %>%
    filter(gt =="1/1") %>%
    group_by(sex) %>% 
    mutate(sites = sum(total)) %>% 
   mutate(count_sample = n_distinct(sample)) %>%
   select(c(sex,sites,count_sample)) %>% 
    unique()

pooled_obs_count

p_01_11_bar = samples_obs_01_11 %>% dplyr::filter(total>=10) %>%
    ggplot(aes(fill=gt, y=prop, x=sex)) + 
    geom_bar(position="fill", stat="identity", ) +
    labs(fill = "Genotype") + 
    theme_classic() +
    theme(axis.text.x = element_text(size=12),
        axis.text.y = element_text(size=12)) +
     scale_fill_manual(values=c("#ffbf00", "#66b032","#1982c4"))+
     theme(legend.position = "none")+
  ggtitle("F0 female, (0/1); F1 male (0/1); F1 female (1/1)")

# histogram plot (count)
p_01_11_hist = samples_obs_01_11 %>% 
  dplyr::filter(total>=10) %>%  
  dplyr::filter(gt == "1/1") %>%
  ggplot(aes(prop, fill = sex)) +
    geom_histogram(binwidth=.05, alpha=.7, position="identity") + theme_light() +
  ggtitle("F1 cross: male (0/1) x female (1/1)",
          subtitle = paste("F2 female = ", filter(pooled_obs_count, sex == 'female')$count_sample, ", male =", filter(pooled_obs_count, sex == 'male')$count_sample))

# save as csv to send to sasha
#write_csv(samples_obs_01_11, "/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/data/samples_prop_01_11.csv")
```

## can we see two seperate populations of F2 males? based on their genotype

- we plot the distribution of F2 males, based on the proportion of sites that behave automicticly (that is, have the same genotype as their mother (F1 female)).  

Do we see two peeks? (that is, can we objectively detect 2 separate populations?  

- If yes –    
we can pick the % threshold that separate the 2 populations, and call them “normal” and “abnormal”.  

- If no –   
We need to keep all males, and filter by site.   

#### Proportion of maternal genotype proportion in F2 mites

```{r}
legend <- get_legend(p_01_11_hist)   # get the legend of the first one plot

# now add the title
#title <- ggdraw() + 
 # draw_label("Maternal genotype proportion in F2 mites", x = 0, y = 3) +
  #theme(plot.margin = margin(0, 0, 0, 7))

prow <- plot_grid(p_00_01_hist + theme(legend.position="none")+scale_x_continuous( n.breaks = 5),
           # here you add the percentage
           p_11_01_hist + theme(legend.position="none")+ scale_x_continuous(n.breaks = 5),
           p_01_00_hist + theme(legend.position="none")+ scale_x_continuous(n.breaks = 5),
          p_01_11_hist + theme(legend.position="none")+ scale_x_continuous(n.breaks = 5),
         align = 'v',
           hjust = -1,
           nrow = 4)

# here you add the legend
plot_grid( prow, legend, rel_widths = c(0.9, 0.4))
``` 

another option:
pool all sites together
```{r eval=FALSE, include=FALSE}
F2_male = rbind(mutate(samples_obs_00_01, cross = "F2_00_01"),
                mutate(samples_obs_11_01, cross = "F2_11_01"), 
                mutate(samples_obs_01_00, cross = "F2_01_00"),
                mutate(samples_obs_01_11, cross = "F2_01_11")) 

pooled_obs_count =  F2_male %>% 
    dplyr::filter(total>=10) %>%
    filter(gt =="0/0") %>%
    group_by(sex) %>% 
    mutate(sites = sum(total)) %>% 
   mutate(count_sample = n_distinct(sample)) %>%
   select(c(sex,sites,count_sample)) %>% 
    unique()

# pool crosses where mother (F1) in hetero
F2_male %>% 
  dplyr::filter(total>=10) %>%  
  dplyr::filter(cross %in% c("F2_00_01","F2_11_01")) %>%
  dplyr::filter(gt == "0/1") %>%
  ggplot(aes(prop, fill = sex)) +
    geom_histogram(binwidth=.05, alpha=.8, position="identity") + theme_light() +
  ggtitle("F2 maternal genotype prop, F1 female 0/1",
          subtitle = paste("F2 female = ", filter(pooled_obs_count, sex == 'female')$count_sample, ", male =", filter(pooled_obs_count, sex == 'male')$count_sample))

# plot all 4 crosses together:
# only males
F2_male %>% 
    dplyr::filter(sex=="male") %>%  
  dplyr::filter(total>=10) %>%  
#  dplyr::filter(parent %in% c("maternal","paternal")) %>% #include both parental origins
  dplyr::filter(parent == "maternal") %>%
 ggplot(aes(prop, fill = parent)) +
    geom_histogram(binwidth=.04, alpha=.8, position="identity", fill = "blue") + theme_light() +
  ggtitle("Proportion of maternal genotype
in F2 males",
          subtitle = paste("Samples = ", filter(pooled_obs_count, sex == 'male')$count_sample, ", sites =", filter(pooled_obs_count, sex == 'male')$sites)) +
        xlab("Site's proportion") + 
      ylab("Count") +
    theme(legend.position="none") +
   geom_vline(xintercept = 0.75, color = "red")


# both males and females
F2_male %>% 
  #  dplyr::filter(sex=="male") %>%  
  dplyr::filter(total>=10) %>%  
  dplyr::filter(parent == "maternal") %>%
 ggplot(aes(prop, fill = sex)) +
    geom_histogram(binwidth=.04, alpha=.7, position="identity") + theme_light() +
  ggtitle("Proportion of maternal genotype
in F2 mites",
          subtitle = paste("Females = ", filter(pooled_obs_count, sex == 'female')$count_sample, ", Males =", filter(pooled_obs_count, sex == 'male')$count_sample)) +
        xlab("Site's proportion") + 
      ylab("Count") +
    theme(legend.position="bottom") +
     scale_fill_discrete(name = "Sex:", labels = c("Female", "Male"))
```

### Figure 2A. Proportion of maternal genotype in F2 mites
```{r}
# both sex, but wrap
F2_male %>% 
  mutate(sex = factor(sex, levels = c("male", "female"))) %>%
  dplyr::filter(total>=10) %>%  
  dplyr::filter(parent == "maternal") %>%
 ggplot(aes(prop)) +
   geom_histogram(aes( fill=sex), binwidth = 0.04, color = "black") +
 theme_classic2() +  
  ggtitle("Proportion of maternal genotype
in F2 mites",
          subtitle = paste("Females = ", filter(pooled_obs_count, sex == 'female')$count_sample, ", Males =", filter(pooled_obs_count, sex == 'male')$count_sample)) +
        xlab("Maternal genotype proportion") + 
      ylab("F2 Mite count") +
      theme(legend.position="none") +
 facet_wrap(~sex, nrow = 2, scales = "free", labeller = labeller(sex = c(`female` = "Females",`male` = "Males"))) +scale_fill_manual(values=c("black","white")) 

```

### who are the abnormal males? set the threshold for maternal genotype proportion to 0.75
```{r}
F2_male %>%  
  dplyr::filter(sex=="male") %>%  
  dplyr::filter(total>=10) %>%  
  dplyr::filter(parent == "maternal") %>% 
  dplyr::filter(prop<=0.75) 

# median, quantiles and avarage 
F2_male %>%
  dplyr::filter(sex=="male") %>%  
   dplyr::filter(total>=10) %>%  
  dplyr::filter(parent == "maternal") %>% stat.desc()

F2_male %>%
  dplyr::filter(sex=="female") %>%  
  dplyr::filter(total>=10) %>%  
  dplyr::filter(parent == "maternal") %>% stat.desc()

# check normality:
fem_maternal_prop = F2_male %>%
  dplyr::filter(sex=="female") %>%  
  dplyr::filter(total>=10) %>%  
  dplyr::filter(parent == "maternal")

  shapiro.test(fem_maternal_prop$prop)
  
  male_maternal_prop = F2_male %>%
  dplyr::filter(sex=="male") %>%  
  dplyr::filter(total>=10) %>%  
  dplyr::filter(parent == "maternal")

  shapiro.test(male_maternal_prop$prop)

#write_csv(F2_male, "/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/data/F2_male_gt_prop.csv")
```
