---
title: "explore VCF statistics using vcftools and vcfR"
author: "Nurit Eliash"
date: "6/17/2021"
output:
  html_document:
    code_folding: hide
    theme: cerulean
    toc: yes
    toc_depth: 5
    toc_float:
      collapsed: yes
      df_print: paged
editor_options: 
  chunk_output_type: console
---
<style type="text/css">
  body{
  font-size: 12pt;
}
</style>

In the following analysis we gonna explore the unfiltered vcf using vcftools, to get an idea what the non-filtered data looks like. 
Based on the results, we can make an informative decision about the actual filtering parameters we wish to use. 
For the exploration, we will use a 0.0015 subset of the original vcf (`snponly_freebayes.vcf`). We also include only variants on the 7 chromosomes, as we are interested in hybridization and crossing over events on the chromosomes. These will reduce the running time of the analysis.

We followed the [Speciation & Population Genomics: a how-to-guide](https://speciationgenomics.github.io/filtering_vcfs/) by Joana Meier and Mark Ravinet.  
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### load libraries
```{r load library, echo = T, results = 'hide', message=FALSE, warning=FALSE}
library(tidyverse)
library(dplyr)
library(ggplot2)
library("vcfR") # for extracting genotype data from a vcf file
library(pinfsc50)
library(reshape)
library(reshape2)
library(hrbrthemes)
library(viridis)
library("plotly") # for the 3d plots
library("gridExtra")
library(cowplot)
library(egg)
```

Set directory
```{r load library, echo = T, results = 'hide', message=FALSE, warning=FALSE}
setwd("/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/data/vcf_filter/7chr_subset")
```

## Generating statistics from a VCF
The statistic files can be found in ***data DIR***. These files were generated using vcftools [Danecek et al. 2011](https://academic.oup.com/bioinformatics/article/27/15/2156/402296), and the scripts can be found in ***scripts DIR***.

## Variant based statistics
The first thing we will do is look at the statistics we generated for each of the variants in our subset VCF - quality, depth, missingness and allele frequency.

### Variant quality
The first metric we will look at is the (Phred encoded) site quality. This is a measure of how much confidence we have in our variant calls. First of all, we read in the site quality report we generated using vcftools. We will use the read_delim command from the readr package (part of the the tidyverse) because it is more efficient for reading in large datafiles. It also allows us to set our own column names.

Take a look at the data when it is read in. You will see that for each site in our subsampled VCF, we have extracted the site quality score. Now we will plot the distribution of this quality using ggplot. Usually, the geom_density function works best, but you can use geom_histogram too.
```{r}
var_qual <- read_delim("/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/data/vcf_filter/7chr_subset/subset_Chr7.lqual", delim = "\t",
           col_names = c("chr", "pos", "qual"), skip = 1)

# make more strict filter on the quality value:
med_fil <- filter(var_qual, qual > 1000)
harsh_fil <- filter(var_qual, qual > 10000)
var_qual_abov1 <- filter(var_qual, qual > 1)

# plot it
a <- ggplot(var_qual, aes(qual)) + geom_density(fill = "dodgerblue1", colour = "black", alpha = 0.3)
a + theme_light() 
  #expand_limits(x=c(20,10000)) 
 
b <- ggplot(med_fil, aes(qual)) + geom_density(fill = "dodgerblue1", colour = "black", alpha = 0.3)
b + theme_light() 

c <- ggplot(harsh_fil, aes(qual)) + geom_density(fill = "dodgerblue1", colour = "black", alpha = 0.3)
c + theme_light() 

d <- ggplot(var_qual_abov1, aes(qual)) + geom_density(fill = "dodgerblue1", colour = "black", alpha = 0.3)
d + theme_light()
summary(var_qual$qual)
```

### Variant mean depth
Next we will examine the mean depth for each of our variants. This is essentially the number of reads that have mapped to this position. The output we generated with vcftools is the mean of the read depth across all individuals - it is for both alleles at a position and is not partitioned between the reference and the alternative. First we read in the data.
```{r}
var_depth <- read_delim("/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/data/vcf_filter/7chr_subset/subset_Chr7.ldepth.mean", delim = "\t",
           col_names = c("chr", "pos", "mean_depth", "var_depth"), skip = 1)

var_depth_15 <- filter(var_depth, mean_depth > 15)
  
a <- ggplot(var_depth, aes(mean_depth)) + geom_density(fill = "dodgerblue1", colour = "black", alpha = 0.3)
a + theme_light()
a + theme_light() + xlim(0, 300)

summary(var_depth$mean_depth)

b <- ggplot(var_depth_15, aes(mean_depth)) + geom_density(fill = "dodgerblue1", colour = "black", alpha = 0.3)
b + theme_light()
```
Again take a moment to look at the data - mean_depth is our column of interest but note that you can also get a an idea of the variance in depth among individuals from the var_depth column. Once again, we will use ggplot to look at the distribution of read depths

### Variant missingness
Next up we will look at the proportion of missingness at each variant. This is a measure of how many individuals lack a genotype at a call site.
```{r}
var_miss <- read_delim("/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/data/vcf_filter/7chr_subset/subset_Chr7.lmiss", delim = "\t",
                       col_names = c("chr", "pos", "nchr", "nfiltered", "nmiss", "fmiss"), skip = 1)

a <- ggplot(var_miss, aes(fmiss)) + geom_density(fill = "dodgerblue1", colour = "black", alpha = 0.3)
a + theme_light()
summary(var_miss$fmiss)
```
Most sites have almost no issing data. Although clearly, there are sum (as the max value shows). This means we can be quite conservative when we set our missing data threshold. 

### Minor allele frequency
Last of all for our per variant analyses, we will take a look at the distribution of allele frequencies. This will help inform our minor-allele frequency (MAF) thresholds. 
```{r}
var_freq <- read_delim("/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/data/vcf_filter/7chr_subset/subset_Chr7.frq", delim = "\t",
                       col_names = c("chr", "pos", "nalleles", "nchr", "a1", "a2"), skip = 1)

#However, this is simply the allele frequencies. To find the minor allele frequency at each site, we need to use a bit of dplyr based code.
# find minor allele frequency
var_freq$maf <- var_freq %>% select(a1, a2) %>% apply(1, function(z) min(z))

# Here we used apply on our allele frequencies to return the lowest allele frequency at each variant. We then added these to our dataframe as the variable maf. Next we will plot the distribution.
frq <- ggplot(var_freq, aes(maf)) + geom_density(fill = "dodgerblue1", colour = "black", alpha = 0.3)
frq + theme_light()

summary(var_freq$maf)

#  excluding --maf 0.2 :
var_freq_0.2 <- filter(var_freq, var_freq$maf > 0.2)
frq.2 <- ggplot(var_freq_0.2, aes(maf)) + geom_density(fill = "dodgerblue1", colour = "black", alpha = 0.3)
frq.2 + theme_light()

#  excluding --maf 0.1 :
var_freq_0.1 <- filter(var_freq, var_freq$maf > 0.1)
frq.1 <- ggplot(var_freq_0.1, aes(maf)) + geom_density(fill = "dodgerblue1", colour = "black", alpha = 0.3)
frq.1 + theme_light()

#  excluding --maf 0.05 :
var_freq_0.05 <- filter(var_freq, var_freq$maf > 0.05)
frq.05 <- ggplot(var_freq_0.05, aes(maf)) + geom_density(fill = "dodgerblue1", colour = "black", alpha = 0.3)
frq.05 + theme_light()
```


###  Allele balance. 
Load Variant Call Format (VCF) file.
I followed this tutorial: [Determining ploidy 1](https://knausb.github.io/vcfR_documentation/determining_ploidy_1.html).  

Extract genotypes for each site and individual. The metadata for all samples can be found in [here](https://github.com/nurit-eliash/varroa-linkage-map/blob/main/data/meta_data_223.xlsx).

```{r}
vcf <- read.vcfR("/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/data/vcf_filter/Q40BIALLDP16HDP40mis.5Chr7/Q40BIALLDP16HDP40mis.5Chr7.recode.vcf", verbose = FALSE )
vcf

knitr::kable(vcf@gt[c(1:2,11,30),1:3])

# The ‘AD’ field in our VCF data includes the depth at which each allele was sequenced. We can extract this information with the function extract.gt().
ad <- extract.gt(vcf, element = 'AD')
knitr::kable(ad[c(1:2,11,30),1:6])

#The function extract.gt() isolates the ‘AD’ data from the colon delimited string in the VCF data. We expect integer counts for the number of sequences observed. However, because this data is comma delimited we need another step before we have integers. We use the function masplit() to extract the first and second allele. At that point we have integers so we can use math to create allele frequencies from the counts.
allele1 <- masplit(ad, record = 1)
allele2 <- masplit(ad, record = 2)

ad1 <- allele1 / (allele1 + allele2)
ad2 <- allele2 / (allele1 + allele2)

#Once we have our allele frequencies we can plot them with a histogram.

hist(ad2[,"240_241c_grnson_S74"], breaks = seq(0,1,by=0.02), col = "#1f78b4", xaxt="n")
hist(ad1[,"240_241c_grnson_S74"], breaks = seq(0,1,by=0.02), col = "#a6cee3", add = TRUE)
axis(side=1, at=c(0,0.25,0.333,0.5,0.666,0.75,1), labels=c(0,"1/4","1/3","1/2","1/3","3/4",1))

#The most common class of variant is the homozygote. This overwhelms the plot. We can remove these so we can focus on the heterozygotes.
gt <- extract.gt(vcf, element = 'GT')
hets <- is_het(gt)

is.na( ad[ !hets ] ) <- TRUE

allele1 <- masplit(ad, record = 1)
allele2 <- masplit(ad, record = 2)

ad1 <- allele1 / (allele1 + allele2)
ad2 <- allele2 / (allele1 + allele2)

hist(ad2[,"476_477d_grndat_S162"], breaks = seq(0,1,by=0.02), col = "#1f78b4", xaxt="n")
hist(ad1[,"476_477d_grndat_S162"], breaks = seq(0,1,by=0.02), col = "#a6cee3", add = TRUE)
axis(side=1, at=c(0,0.25,0.333,0.5,0.666,0.75,1), labels=c(0,"1/4","1/3","1/2","1/3","3/4",1))

#This allows us to focus on the heterozygous variants and we observe a peak at 0.5 consistent with our expectation. However, the frequencies range almost completely from 0 to one. This suggests some improvement could be made.

# you can play with the differnt samples.

# "normal" males, excepted parthenogenetic inheritance: 476_477b_grnson_S160, 284_285a_grnson_S85, 600_601a_grnson_S153
# "weird" males, exceptional inheritance:  400_401a_grnson_S112, 412_413a_grnson_S116, 458_459a_grnson_S130, 46_47d_grnson_S16, 534_535_2c_grnson_S141, 426_427b_grnson_S122

## females:
# normal females:  476_477a_grndat_S159,476_477d_grndat_S162, 478_479-1b_grndat_S168, 338_339a_grndat_S104, 498_499c_grndat_S180, 43_44a_grndat_S5, 534_535_2a_grndat_S139
#wired females:   240_241a_grndat_S73, 240_241b_grndat_S133
# too little sites.. : 564_565-2a_grndat_S206, 177_178a_grndat_S60, 478_479-2d_grndat_S175
```

### Plot all samples Allele balance
```{r}
ad1M <- reshape::melt(ad1)
colnames(ad1M) = c("site", "sample", "value")
#clean the sample names
ad1M$sample <- sub("_[^_]+$", "", ad1M$sample)
ad1M = ad1M %>%
  mutate(allele = "ref")

ad2M <- reshape::melt(ad2)
colnames(ad2M) = c("site", "sample", "value")
#clean the sample names
ad2M$sample <- sub("_[^_]+$", "", ad2M$sample) 
ad2M = ad2M %>%
  mutate(allele = "alt")

# add the sex of each sample, and the proportion of homosyzgotic and hetero sites
info <- read.csv("/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/data/meta_data_223.csv") %>%
  select(!"sample")
colnames(info) <- c("sample","family","dev_stage",  "generation" ,"sex","pedigree"  )

table = bind_rows(ad1M, ad2M) %>%
 replace(is.na(.),0) %>% 
  left_join(info, by = "sample")

#plot adult males
P_male_no0_1 = table %>%  dplyr::filter(sex == "male") %>%
  filter(value>0) %>% # remove the very much homozygotic sites
  filter(value<1) %>%
  ggplot( aes(x=value, fill= allele)) +
    geom_histogram(color="black", position = 'identity') +
        theme_classic() +
        theme(
      legend.position="none",
      panel.spacing = unit(0.1, "lines"),
      strip.text.x = element_text(size = 8)) +
    xlab("Allele balance") +
    ylab("Site count")+
  ggtitle("Males Allele balance, no 'zeros' and 'ones'") +
  theme(legend.position="bottom")  +
    facet_wrap(~sample, scales = "free") 

P_male_includ_homo = table %>%   dplyr::filter(sex == "male")  %>%
 # filter(value>0) %>% 
#  filter(value<1) %>%
  ggplot( aes(x=value, fill= allele)) +
    geom_histogram(color="black", position = 'identity') +
        theme_classic() +
        theme(
      legend.position="none",
      panel.spacing = unit(0.1, "lines"),
      strip.text.x = element_text(size = 8)) +
    xlab("Allele balance") +
    ylab("Site count")+
  ggtitle("Males Allele balance, keep all homo sites") +
  theme(legend.position="bottom")  +
    facet_wrap(~sample, scales = "free") 

#plot adult females
P_fem_no0_1 = table %>%   dplyr::filter(sex == "female")  %>%
  filter(value>0) %>% # remove the very much homozygotic sites
  filter(value<1) %>%
  ggplot( aes(x=value, fill= allele)) +
    geom_histogram(color="black", position = 'identity') +
        theme_classic() +
        theme(
      legend.position="none",
      panel.spacing = unit(0.1, "lines"),
      strip.text.x = element_text(size = 8)) +
    xlab("Allele balance") +
    ylab("Site count")+
  ggtitle("Female Allele balance, no 'zeros' and 'ones'") +
  theme(legend.position="bottom")  +
    facet_wrap(~sample, scales = "free") 

P_fem_includ_homo = table %>%   dplyr::filter(sex == "female") %>%
 filter(value>0) %>% 
 filter(value<1) %>%
  ggplot( aes(x=value, fill= allele)) +
    geom_histogram(color="black", position = 'identity') +
        theme_classic() +
        theme(
      legend.position="none",
      panel.spacing = unit(0.1, "lines"),
      strip.text.x = element_text(size = 8)) +
    xlab("Allele balance") +
    ylab("Site count")+
  ggtitle("Females Allele balance, keep all homo sites") +
  theme(legend.position="bottom")  +
    facet_wrap(~sample, scales = "free") 
```

#### allele balance by family
```{r}
#families = c("240","63")
families = table$family %>% unique() %>% as.character()
plotlist = list()

# looping over families in table
for(fam in families) {
 plotlist[[fam]] =
  table %>% filter(family == fam) %>%
    filter(value>0) %>% 
 filter(value<1) %>%
  ggplot( aes(x=value, fill= allele)) +
    geom_histogram(color="black", position = 'identity') +
        theme_classic() +
        theme(
      legend.position="none",
      panel.spacing = unit(0.1, "lines"),
      strip.text.x = element_text(size = 8)) +
    xlab("Allele balance") +
    ylab("Site count")+
  ggtitle(paste0("Family ", fam)) +
  theme(legend.position="blank")  +
    facet_wrap(~sample, scales = "free", nrow=1) 
}

all_plots = grid.arrange(grobs = plotlist, nrow=30, egg::set_panel_size)

ggsave("/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/results/plots/myplot.pdf", plot = all_plots,width = 40, height = 100, units = "cm")





####

### for 2 plots:
families = c("240","400","63")
plotlist = list()

# looping over families in table
for(fam in families) {
 plotlist[[fam]] =
  table %>% filter(family == fam) %>%
    filter(value>0) %>% 
 filter(value<1) %>%
  ggplot( aes(x=value, fill= allele)) +
    geom_histogram(color="black", position = 'identity') +
        theme_classic() +
        theme(
      legend.position="none",
      panel.spacing = unit(0.1, "lines"),
      strip.text.x = element_text(size = 8)) +
    xlab("Allele balance") +
    ylab("Site count")+
  ggtitle(paste0("Family ", fam)) +
  theme(legend.position="blank")  +
    facet_wrap(~sample, scales = "free", nrow=1) 
}


all_plots = grid.arrange(grobs = plotlist, nrow=3,  common.legend = TRUE, legend="bottom", widths = unit(0.9, "npc"),  align = "v")

#position.units = "npc", size.units = "mm"), heights = unit(0.5, "npc")

ggsave("/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/results/plots/myplot.pdf", plot = all_plots,width = 40, height = 20, units = "cm")


#### זה מאוד מעצבן
# try to find a way to force all plots to be the same size , regardless of number of plots in a facet

```


## Genetic differentiation
measure population diversity and differentiation. Because VCF data typically do not include population information we’ll have to supply it as a factor. The method ‘nei’ employed here is based on the methods reported by Hedrick (Hedrick 2005). The exception is that the heterozygosities are weighted by the number of alleles observed in each population. This was inspired by hierfstat::pairwise.fst() which uses the number of individuals observed in each population to weight the heterozygosities. By using the number of alleles observed instead of the number of individuals we remove an assumption about how many alleles each individual may contribute. That is, we should be able to accomodate samples of mixed ploidy. you can find the tutorial in [here](https://knausb.github.io/vcfR_documentation/genetic_differentiation.html)
```{r}
pop <- as.factor(c("us", "eu", "us", "af", "eu", "us", "mx", "eu", "eu", "sa", "mx", "sa", "us", "sa", "Pmir", "us", "eu", "eu"))
myDiff <- genetic_diff(vcf, pops = pop, method = 'nei')
knitr::kable(head(myDiff[,1:15]))
knitr::kable(head(myDiff[,16:19]))

#We now have information for each variant in the VCF data. Because this is typically a large quantity of information, we’ll want to summarize it. One way is to take averages of the data.
knitr::kable(round(colMeans(myDiff[,c(3:9,16,19)], na.rm = TRUE), digits = 3))

#Another way to summarize data is to use violin plots.
dpf <- reshape::melt(myDiff[,c(3:8,19)], varnames=c('Index', 'Sample'), value.name = 'Depth', na.rm=TRUE)
p <- ggplot(dpf, aes(x=variable, y=Depth)) + 
  geom_violin(fill="#2ca25f", adjust = 1.2) + 
  xlab("") + 
  theme_bw()
```

## Individual based statistics
As well as a our per variant statistics we generated earlier, we also calculated some individual metrics too. WE can look at the distribution of these to get an idea whether some of our individuals have not sequenced or mapped as well as others. This is good practice to do with a new dataset. A lot of these statistics can be compared to other measures generated from the data (i.e. principal components as a measure of population structure) to see if they drive any apparent patterns in the data.

### Mean depth per individual
First we will look at the distribution of mean depth among individuals. We read the data in with read_delim:
```{r}
# NO filtering
ind_depth <- read_delim("//Users/nuriteliash/Documents/GitHub/varroa-linkage-map/data/vcf_filter/7chr_subset/subset_Chr7.idepth", delim = "\t",
                        col_names = c("ind", "nsites", "depth"), skip = 1)

#Then we plot the distribution as a histogram using ggplot and geom_hist.
a <- ggplot(ind_depth, aes(depth)) + geom_histogram(fill = "dodgerblue1", colour = "black", alpha = 0.3)
a + theme_light()

a <- ggplot(ind_depth, aes(depth)) + geom_density(fill = "dodgerblue1", colour = "black", alpha = 0.3)
a + theme_light()

summary(ind_depth$depth)
```

### Proportion of missing data per individual
Next we will look at the proportion of missing data per individual. We read in the data below:
```{r}
ind_miss  <- read_delim("/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/data/vcf_filter/7chr_subset/subset_Chr7.imiss", delim = "\t",
                        col_names = c("ind", "ndata", "nfiltered", "nmiss", "fmiss"), skip = 1)
a <- ggplot(ind_miss, aes(fmiss)) + geom_histogram(fill = "dodgerblue1", colour = "black", alpha = 0.3)
a + theme_light()

a <- ggplot(ind_miss, aes(fmiss)) + geom_density(fill = "dodgerblue1", colour = "black", alpha = 0.3)
a + theme_light()

summary(ind_miss$fmiss)
```

### Heterozygosity and inbreeding coefficient per individual
```{r}
ind_het <- read_delim("/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/data/vcf_filter/Q40BIALLDP16HDP40mis.5Chr7/Q40BIALLDP16HDP40mis.5Chr7.het", delim = "\t", col_names = c("sample","ho", "he", "nsites", "f"), skip = 1)

# add the sex of each sample, and the proportion of homosyzgotic and hetero sites
meta <- read.csv("/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/data/meta_data_223.csv")
ind_het <- left_join(ind_het,meta, by ="sample") %>%
  mutate(hom_prop =  ho/nsites) %>%
  mutate(het_prop = (nsites-ho)/nsites) %>%
mutate_at(c("family", "generation"), as.character)
ind_het$sample <- sub("_[^_]+$", "", ind_het$sample)

#plot inbreeding coefficient per individual
ggplot(ind_het, aes(f, fill = sex)) + 
  geom_histogram(color = "black", position = 'identity') + 
  theme_light() +
  ggtitle("Inbreeding coefficient per individual")

# plot heterozygosity proportion, in each sex
ind_het %>% filter(sex== c("male","female")) %>%
  ggplot() +
    geom_boxplot(aes(x = sex, y = het_prop, fill = sex)) + scale_y_continuous() + 
    theme_classic() +
    ggtitle("Proportion of heterzygotic sites in varroa genome") +
      scale_y_continuous(expand=c(0,0), limits = c(0, 1))

p_het = ind_het %>% filter(sex== c("male","female")) %>%
  ggplot(aes(x=family, y=het_prop, color=sex, label = pedigree)) + 
    geom_point(size=1) +
    #scale_shape_manual(values=c(1,2,3)) +
scale_y_continuous() + 
    theme_classic() +
    ggtitle("Proportion of heterzygotic sites in varroa genome") + geom_text(hjust=0, vjust=0)

 # facet_wrap(~sex)

ggplotly(p_het)
# from this analysis it looks like nothing of the observed factors has an effect on the mite heterozygosity:
# the heterozygosity is similar in males and females, not gained/decreased over generations (F0, F1 and F2), 

# the only thing i can tell is, that there are specific families with higher heterozygosity levels
# and that the hetero level is similar in females and their son (F0 female ~ F1 male) and (F1 female ~ F2 male), within each family. this actually makes sense, assuming parthenogenesis. 

ind_het %>% filter(sex== c("male","female")) %>%
  ggplot() +
    geom_boxplot(aes(x = sex, y = het_prop, fill = sex)) + scale_y_continuous() + 
    theme_classic() +
    ggtitle("Proportion of heterzygotic sites in varroa genome") 

means= ind_het %>% filter(sex== c("male","female")) %>%
group_by(sex, ) %>%                        
summarise_at(vars(het_prop),
              list(name = mean))

```
