---
title: "variant_position"
author: "Nurit Eliash"
date: "6/4/2022"
output: html_document
---

```{r setup, include=FALSE}
setwd("/Users/nuriteliash/Desktop/")
library("RIdeogram")
library("tidyverse")
library("plyr")
library("dplyr")
library("ggplot2")
library("scales")
library("ggpubr")
library("gridExtra")
library("grid")
library("GGally")
library("vcfR") # for extracting genotype data from a vcf file
library("data.table")
library("stringr")
library("janitor")
knitr::opts_chunk$set(echo = TRUE)
```

## Load Variant Call Format (VCF) file.
Extract genotypes for each site and individual
```{r}
vcf <- read.vcfR("/Users/nuriteliash/Documents/GitHub/varroa-linkage-map/data/vcf_filter/Q40BIALLDP16HDP40mis.5Chr7/Q40BIALLDP16HDP40mis.5Chr7.recode.vcf", verbose = FALSE )
vcf
# extract the genotype for each site in each individual
gt <- extract.gt(vcf, element = "GT") 
gt <- as.data.frame(t(gt)) %>%
  rownames_to_column("ID")
#clean the ID names
gt$ID <- sub("_[^_]+$", "", gt$ID)

table <-  gt %>% 
  t() %>%
  as.data.frame() %>%
  row_to_names(row_number = 1) %>%
  dplyr::select(contains(c("son", "dat"))) # keep only adults of F1 and F2 
```

make the list of sites in wch genotype for males and females
### (5) F1 male (0/0) x female (0/1)
```{r}
# set the families (include only families with at least one adult F2)
family = grep("grndat|grnson",gt$ID, value=TRUE) %>%
  str_extract("[^_]+")  %>%
  unique()

# define a list to put all the data frames in
list <- list()

for (fam in family) {
 
list[[fam]] <- table %>%
  dplyr::select(starts_with(fam)) %>%
  dplyr::filter_at(vars(matches("_son")), all_vars(. == "0/0")) %>%
  dplyr::filter_at(vars(matches("_dat")), all_vars(. == "0/1")) %>% 
  dplyr::select(contains("grn")) %>% rownames_to_column("site") 
}

# join all grandkids of 26 families together
F2all <- reduce(list, full_join, by = "site")

# separate to male and females of the F2 
F2male <- F2all %>%
  dplyr::select(dplyr::contains(c("site","son"))) %>% mutate_all(as.character) %>%
  mutate("count0/0" = rowSums(F2male[-1] == "0/0", na.rm=TRUE)) %>%
  mutate("count0/1" = rowSums(F2male[-1] == "0/1", na.rm=TRUE)) %>%
  mutate("count1/1" = rowSums(F2male[-1] == "1/1", na.rm=TRUE)) %>%
  separate(col="site", into=c("noting","Chr" ,"Start"), sep="_") %>% 
  mutate(Start = as.double(Start)) %>%
  mutate("End"=Start+3) %>%
  dplyr::select(c("Chr","Start","End","count0/0","count0/1","count1/1"))

F2female <- F2all %>% 
  dplyr::select(dplyr::contains(c("site","dat"))) %>% mutate_all(as.character) %>%
  mutate("count0/0" = rowSums(F2female[-1] == "0/0", na.rm=TRUE)) %>%
  mutate("count0/1" = rowSums(F2female[-1] == "0/1", na.rm=TRUE)) %>%
  mutate("count1/1" = rowSums(F2female[-1] == "1/1", na.rm=TRUE)) %>%
  separate(col="site", into=c("noting","Chr" ,"Start"), sep="_") %>% 
  mutate(Start = as.double(Start)) %>%
  mutate("End"=Start+3) %>%
  dplyr::select(c("Chr","Start","End","count0/0","count0/1","count1/1"))
```

make a table with chromosome info, and plot their position
```{r eval=FALSE, include=FALSE}
# make the karyotype table with chromosomes start and end position
karyotype <- tibble(Chr = c("NW_019211454.1", "NW_019211455.1", "NW_019211456.1", "NW_019211457.1", "NW_019211458.1", "NW_019211459.1", "NW_019211460.1"),
           Start = 0,
             End = c(76960006, 60513814,58583513,52932055,42024542,32556157,39431147),
           ) %>% separate(col=Chr, into=c("noting","Chr"), sep="_") %>%
  dplyr::select(-"noting")
  
sites <- F2male %>%
    dplyr::select(c("Chr","Start","End","count0/1")) %>%
    dplyr::rename("Value" = "count0/1")

#ideogram(karyotype = karyotype)
ideogram(karyotype = karyotype, overlaid = sites)
convertSVG("chromosome.svg", device = "png")
```

#########

# per family?
make the list of sites in wch genotype for males and females
### (5) F1 male (0/0) x female (0/1)
```{r}
# set the families (include only families with at least one adult F2)
family = grep("grndat|grnson",gt$ID, value=TRUE) %>%
  str_extract("[^_]+")  %>%
  unique()

# define a list to put all the data frames in
list <- list()

for (fam in family) {
 
list[[fam]] <- table %>%
  dplyr::select(starts_with(fam)) %>%
  dplyr::filter_at(vars(matches("_son")), all_vars(. == "0/0")) %>%
  dplyr::filter_at(vars(matches("_dat")), all_vars(. == "0/1")) %>% 
  rownames_to_column("site") 
}

# join all grandkids of 26 families together
all0.0_0.1 <- reduce(list, full_join, by = "site")

# color discrete - different color for each genotype 
color_gt = tibble(Value = as.numeric(c("0","1","2")),
                  gt = c("0/1","0/0","1/1"))

# compare F2 male and female, and their F1 parents for one family
sites_gt <- all0.0_0.1 %>%
  dplyr::select(dplyr::contains(c("site","284_284a_son"))) %>% 
  separate(col=site, into=c("noting","Chr" ,"Start"), sep="_") %>% 
  na.omit() %>%
  mutate(Start = as.double(Start)) %>%
  mutate("End"=Start+50000) %>%
  dplyr::select(c("Chr","Start","End","284_284a_son")) %>%
  dplyr::rename("gt" = "284_284a_son") %>%
  left_join(color_gt, by = "gt") %>%
  dplyr::select(c("Chr","Start","End","Value"))
  
karyotype <- tibble(Chr = c("NW_019211454.1", "NW_019211455.1", "NW_019211456.1", "NW_019211457.1", "NW_019211458.1", "NW_019211459.1", "NW_019211460.1"),
           Start = 0,
             End = c(76960006, 60513814,58583513,52932055,42024542,32556157,39431147),
           ) %>% separate(col=Chr, into=c("noting","Chr"), sep="_") %>%
  dplyr::select(-"noting")

ideogram(karyotype = karyotype, overlaid = sites_gt, colorset1 = c("#49be25","#be4d25","#2596be"))
convertSVG("chromosome.svg", device = "png")


```






# ### old one - per one ID:
# color discrete - different color for each genotype 
color_gt = tibble(Value = as.numeric(c("0","1","2")),
                  gt = c("0/1","0/0","1/1"))

site_gt <- table %>% rownames_to_column("site") %>% 
  dplyr::select(c("site","63_63a_son")) %>% 
  separate(col=site, into=c("noting","Chr" ,"Start"), sep="_") %>% 
  na.omit() %>%
  mutate(Start = as.double(Start)) %>%
  mutate("End"=Start+1) %>%
  dplyr::rename("gt" = "63_63a_son") %>%
  left_join(color_gt, by = "gt") %>%
  dplyr::select(c("Chr","Start","End","Value"))

#ideogram(karyotype = karyotype)
ideogram(karyotype = karyotype, overlaid = site_gt, colorset1 = c("#49be25","#be4d25","#2596be"))


#ideogram(karyotype = karyotype, label = rand_sites, label_type = "marker")
convertSVG("chromosome.svg", device = "png")
```

not in
```{r }

# make the  for one sample: 
sites_genotype <- table %>% rownames_to_column("site") %>% 
  dplyr::select(c("site","63_63a_son")) %>% 
  separate(col=site, into=c("noting","Chr" ,"Start"), sep="_") %>% 
  na.omit() %>%
  mutate(Start = as.double(Start)) %>%
  mutate("End"=Start+1) %>%
  dplyr::select(c("Chr","Start","End","63_63a_son"))

# color discrete - different color for each genotype 
color_gt = tibble(Value = as.numeric(c("0","1","2")),
                  gt = c("0/1","0/0","1/1"))

site_gt <- table %>% rownames_to_column("site") %>% 
  dplyr::select(c("site","63_63a_son")) %>% 
  separate(col=site, into=c("noting","Chr" ,"Start"), sep="_") %>% 
  na.omit() %>%
  mutate(Start = as.double(Start)) %>%
  mutate("End"=Start+1) %>%
  dplyr::rename("gt" = "63_63a_son") %>%
  left_join(color_gt, by = "gt") %>%
  dplyr::select(c("Chr","Start","End","Value"))

ideogram(karyotype = karyotype, overlaid = site_gt, colorset1 = c("#49be25","#be4d25","#2596be"))
convertSVG("chromosome.svg", device = "png")

###########
# didnt use that code:
color_gt = tibble(color = c("#49be25","#be4d25","#2596be"),
                  Type = c("0/1","0/0","1/1"),
                  Shape = c("circle")) 
sites <- table %>% rownames_to_column("site") %>% 
  dplyr::select(c("site","63_63a_son")) %>% 
  separate(col=site, into=c("noting","Chr" ,"Start"), sep="_") %>% 
  na.omit() %>%
  mutate(Start = as.double(Start)) %>%
  mutate("End"=Start+1) %>%
  dplyr::rename("Type" = "63_63a_son") %>%
  left_join(color_gt, by = "Type") %>%
  dplyr::select(c("Type","Shape","Chr","Start","End","color"))

#rand_sites <- sites[sample(nrow(sites), size=150), ]
```

### (9) F1 male (0/1) x female (0/1)
```{r}
# set the families (include only families with at least one adult F2)
family = grep("grndat|grnson",gt$ID, value=TRUE) %>%
  str_extract("[^_]+")  %>%
  unique()

# define a list to put all the data frames in
obs <- list()

for (fam in family) {
 
obs[[fam]] <- table %>%
  dplyr::select(starts_with(fam)) %>%
  dplyr::filter_at(vars(matches("_son")), all_vars(. == "0/1")) %>%
  dplyr::filter_at(vars(matches("_dat")), all_vars(. == "0/1")) %>% 
  dplyr::select(contains("grn")) %>%
  tidyr::pivot_longer(everything())  %>% 
  dplyr::rename(sample = name, gt = value) %>%
  dplyr::count(sample, gt, .drop = FALSE) %>% 
  dplyr::filter(gt %in% c("0/0", "1/1", "0/1")) %>%
  mutate(n = as.numeric(n)) %>%
  group_by(sample) %>%
  mutate(total = as.numeric(sum(n))) %>%
  dplyr::rename(obs = n) %>%
  mutate(sex = case_when(
    grepl("son", sample) ~ "male",
    grepl("dat", sample) ~ "female"))
}

# bind all families together, to a final data frame containing all observed counts
observed <- do.call("rbind", obs) 

# state the expected site proportion of each genotype and sex under the different hypothesis:
# Heterozygotic sites are real:Males can carry and inherit two alleles
fem_sitesReal <- data.frame(type = "sitesReal", sex = "female", gt = c("0/0", "0/1", "1/1") , prop = c(0.25, 0.5, 0.25))

# Heterozygotic sites are false:Males can carry and inherit only one allele
# the heterozygotic sites are actually (0)
fem_sitesFalse_0 <- data.frame(type = "sitesFalse_0",sex = "female", gt = c("0/0", "0/1", "1/1") , prop = c(0.5, 0.5, 0))
# the heterozygotic sites are actually (1)
fem_sitesFalse_1 <- data.frame(type = "sitesFalse_1", sex = "female", gt = c("0/0", "0/1", "1/1") , prop = c(0, 0.5, 0.5))

# males F2 genotype is expected to stay the same , as it is produced from unfertilized egg, hence have no paternal inheritance
male_sitesReal <- data.frame(type = "sitesReal",sex = "male",gt = c("0/0", "0/1", "1/1"),prop = c(0.5, 0, 0.5))
male_sitesFalse_0 <- data.frame(type = "sitesFalse_0",sex = "male",gt = c("0/0", "0/1", "1/1"),prop = c(0.5, 0, 0.5))
male_sitesFalse_1 <- data.frame(type = "sitesFalse_1", sex = "male",gt = c("0/0", "0/1", "1/1"),prop = c(0.5, 0, 0.5))

expected_prop <- bind_rows(fem_sitesReal,fem_sitesFalse_0,fem_sitesFalse_1,male_sitesReal,male_sitesFalse_0,male_sitesFalse_1) %>% mutate(sex = as.character(sex))%>% mutate(gt = as.character(gt)) 
 # mutate(type = "exp")

# using on the total genotypes count of sites per sample, calculate the expected counts, and add them to the observed counts into a final dat table
expected <- observed %>%
  mutate(gt = as.character(gt)) %>%
  full_join(expected_prop, by = c("gt", "sex")) %>% 
  mutate(count = total*prop)

x <- observed %>%
  dplyr::select(c("sample","sex","gt","total", "obs")) %>%
  mutate(type = "Observed") %>%
  dplyr::rename(count = obs) %>%
  mutate(prop = (count/total))

dat <-  bind_rows(x, expected) %>% 
  dplyr::select(-obs)
```

```{r}
# subset for males and females, and reorder the "types" levels:
dat_fem <- dat %>% dplyr::filter(sex == "female")
level_order_fem <- factor(dat_fem$type, level = c("sitesFalse_0", "sitesFalse_1", "sitesReal", "Observed"))

dat_male <- dat %>% dplyr::filter(sex == "male")
level_order_male <- factor(dat_male$type, level = c("sitesFalse_0", "sitesFalse_1", "sitesReal", "Observed"))

p_male_pooled <- ggplot(dat_male, aes(fill=gt, y=prop, x=level_order_male)) + 
    geom_bar(position="fill", stat="identity", width = 0.8) +
    ylab("Proportion of F2 genotype") +
    labs(title = paste0("F2 Male, pooled sites = ", dat_male %>% dplyr::filter(type == "Observed") %>% summarise(sum(count)) %>% as.numeric()))  + 
  labs(fill = "Genotype:") + 
  theme_classic() +
  theme(axis.text.x = element_text(size=12, angle = 45,hjust = 1),
        axis.title.x = element_blank())

p_female_pooled <- ggplot(dat_fem, aes(fill=gt, y=prop, x=level_order_fem)) + 
    geom_bar(position="fill", stat="identity", width = 0.8) +
    ylab("Proportion of F2 genotype") +
    labs(title = paste0("F2 Female, pooled sites = ", dat_fem %>% dplyr::filter(type == "Observed") %>% summarise(sum(count)) %>% as.numeric())) + 
  labs(fill = "Genotype:") +
  theme_classic() +
  theme(axis.text.x = element_text(size=12, angle = 45,hjust = 1),
        axis.title.x = element_blank())

combined <- p_female_pooled + p_male_pooled & theme(legend.position = "bottom")
combined + plot_layout(guides = "collect") + plot_annotation(
  title = 'F2 genotype. F1 Cross: male (0/1) x female (0/1)')

# plot each individual:
ggplot(dat_male, aes(fill=gt, y=prop, x=level_order_male)) + 
    geom_bar(position="fill", stat="identity", ) +
    ylab("Proportion of F2 genotype") +
    labs(title = "F2 male genotype. F1 Cross: male (0/1) x female (0/1)", fill = "Genotype:") +
    geom_text(data = dat_male %>% filter(gt == "0/1") %>% filter(type == "Observed"), aes(x = 4, y = 0.8, label = total), inherit.aes = FALSE) +
    facet_wrap(~ sample) +
    theme_classic()+ 
    theme(legend.position = "bottom", axis.title.x = element_blank(), axis.text.x = element_text(size=8, angle = 45,hjust = 1)) 
        

ggplot(dat_fem, aes(fill=gt, y=prop, x=level_order_fem)) + 
    geom_bar(position="fill", stat="identity", ) +
    ylab("Proportion of F2 genotype") +
    labs(title = "F2 female genotype. F1 Cross: male (0/1) x female (0/1)", fill = "Genotype:") + 
    geom_text(data = dat_fem %>% filter(gt == "0/1") %>% filter(type == "Observed"), aes(x = 4, y = 0.8, label = total), inherit.aes = FALSE) +
    facet_wrap(~ sample) +
    theme_classic()+ 
    theme(legend.position = "bottom", axis.title.x = element_blank(), axis.text.x = element_text(size=8, angle = 45,hjust = 1)) 
```
